// n8n/dist/runners/runner-ws-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupRunnerHandler = exports.setupRunnerServer = exports.TaskRunnerService = void 0;
const config_1 = require("@n8n/config");
const http_1 = require("http");
const n8n_workflow_1 = require("n8n-workflow");
const typedi_1 = __importStar(require("typedi"));
const url_1 = require("url");
const ws_1 = require("ws");
const logger_service_1 = require("../logging/logger.service");
const response_helper_1 = require("../response-helper");
const task_runner_auth_controller_1 = require("../runners/auth/task-runner-auth.controller");
const task_broker_service_1 = require("./task-broker.service");
function heartbeat() {
    this.isAlive = true;
}
function getEndpointBasePath(restEndpoint) {
    const globalConfig = typedi_1.default.get(config_1.GlobalConfig);
    let path = globalConfig.taskRunners.path;
    if (path.startsWith('/')) {
        path = path.slice(1);
    }
    if (path.endsWith('/')) {
        path = path.slice(-1);
    }
    return `/${restEndpoint}/${path}`;
}
function getWsEndpoint(restEndpoint) {
    return `${getEndpointBasePath(restEndpoint)}/_ws`;
}
let TaskRunnerService = class TaskRunnerService {
    constructor(logger, taskBroker) {
        this.logger = logger;
        this.taskBroker = taskBroker;
        this.runnerConnections = new Map();
    }
    sendMessage(id, message) {
        this.runnerConnections.get(id)?.send(JSON.stringify(message));
    }
    add(id, connection) {
        connection.isAlive = true;
        connection.on('pong', heartbeat);
        let isConnected = false;
        const onMessage = (data) => {
            try {
                const buffer = Array.isArray(data) ? Buffer.concat(data) : Buffer.from(data);
                const message = JSON.parse(buffer.toString('utf8'));
                if (!isConnected && message.type !== 'runner:info') {
                    return;
                }
                else if (!isConnected && message.type === 'runner:info') {
                    this.removeConnection(id);
                    isConnected = true;
                    this.runnerConnections.set(id, connection);
                    this.taskBroker.registerRunner({
                        id,
                        taskTypes: message.types,
                        lastSeen: new Date(),
                        name: message.name,
                    }, this.sendMessage.bind(this, id));
                    this.sendMessage(id, { type: 'broker:runnerregistered' });
                    this.logger.info(`Runner "${message.name}"(${id}) has been registered`);
                    return;
                }
                void this.taskBroker.onRunnerMessage(id, message);
            }
            catch (error) {
                this.logger.error(`Couldn't parse message from runner "${id}"`, {
                    error: error,
                    id,
                    data,
                });
            }
        };
        connection.once('close', () => {
            connection.off('pong', heartbeat);
            connection.off('message', onMessage);
            this.removeConnection(id);
        });
        connection.on('message', onMessage);
        connection.send(JSON.stringify({ type: 'broker:inforequest' }));
    }
    removeConnection(id) {
        const connection = this.runnerConnections.get(id);
        if (connection) {
            this.taskBroker.deregisterRunner(id);
            connection.close();
            this.runnerConnections.delete(id);
        }
    }
    handleRequest(req, _res) {
        this.add(req.query.id, req.ws);
    }
};
exports.TaskRunnerService = TaskRunnerService;
exports.TaskRunnerService = TaskRunnerService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [logger_service_1.Logger,
        task_broker_service_1.TaskBroker])
], TaskRunnerService);
const setupRunnerServer = (restEndpoint, server, app) => {
    const globalConfig = typedi_1.default.get(config_1.GlobalConfig);
    const { authToken } = globalConfig.taskRunners;
    if (!authToken) {
        throw new n8n_workflow_1.ApplicationError('Authentication token must be configured when task runners are enabled. Use N8N_RUNNERS_AUTH_TOKEN environment variable to set it.');
    }
    const endpoint = getWsEndpoint(restEndpoint);
    const wsServer = new ws_1.Server({ noServer: true });
    server.on('upgrade', (request, socket, head) => {
        if ((0, url_1.parse)(request.url).pathname !== endpoint) {
            return;
        }
        wsServer.handleUpgrade(request, socket, head, (ws) => {
            request.ws = ws;
            const response = new http_1.ServerResponse(request);
            response.writeHead = (statusCode) => {
                if (statusCode > 200)
                    ws.close();
                return response;
            };
            app.handle(request, response);
        });
    });
};
exports.setupRunnerServer = setupRunnerServer;
const setupRunnerHandler = (restEndpoint, app) => {
    const wsEndpoint = getWsEndpoint(restEndpoint);
    const authEndpoint = `${getEndpointBasePath(restEndpoint)}/auth`;
    const taskRunnerAuthController = typedi_1.default.get(task_runner_auth_controller_1.TaskRunnerAuthController);
    const taskRunnerService = typedi_1.default.get(TaskRunnerService);
    app.use(wsEndpoint, taskRunnerAuthController.authMiddleware, (req, res) => taskRunnerService.handleRequest(req, res));
    app.post(authEndpoint, (0, response_helper_1.send)(async (req) => await taskRunnerAuthController.createGrantToken(req)));
};
exports.setupRunnerHandler = setupRunnerHandler;
