// n8n-nodes-base/dist/credentials/Postgres.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Postgres = void 0;
const sshTunnel_properties_1 = require("../utils/sshTunnel.properties");
class Postgres {
    constructor() {
        this.name = 'postgres';
        this.displayName = 'Postgres';
        this.documentationUrl = 'postgres';
        this.properties = [
            {
                displayName: 'Host',
                name: 'host',
                type: 'string',
                default: 'localhost',
            },
            {
                displayName: 'Database',
                name: 'database',
                type: 'string',
                default: 'postgres',
            },
            {
                displayName: 'User',
                name: 'user',
                type: 'string',
                default: 'postgres',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Ignore SSL Issues',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                default: false,
                description: 'Whether to connect even if SSL certificate validation is not possible',
            },
            {
                displayName: 'SSL',
                name: 'ssl',
                type: 'options',
                displayOptions: {
                    show: {
                        allowUnauthorizedCerts: [false],
                    },
                },
                options: [
                    {
                        name: 'Allow',
                        value: 'allow',
                    },
                    {
                        name: 'Disable',
                        value: 'disable',
                    },
                    {
                        name: 'Require',
                        value: 'require',
                    },
                ],
                default: 'disable',
            },
            {
                displayName: 'Port',
                name: 'port',
                type: 'number',
                default: 5432,
            },
            ...sshTunnel_properties_1.sshTunnelProperties,
        ];
    }
}
exports.Postgres = Postgres;
