// @n8n/task-runner/dist/authenticator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.authenticate = authenticate;
const n8n_workflow_1 = require("n8n-workflow");
const a = __importStar(require("node:assert/strict"));
async function authenticate(opts) {
    try {
        const authEndpoint = `http://${opts.n8nUri}/runners/auth`;
        const response = await fetch(authEndpoint, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                token: opts.authToken,
            }),
        });
        if (!response.ok) {
            throw new n8n_workflow_1.ApplicationError(`Invalid response status ${response.status}: ${await response.text()}`);
        }
        const { data } = (await response.json());
        const grantToken = data.token;
        a.ok(grantToken);
        return grantToken;
    }
    catch (e) {
        console.error(e);
        const error = e;
        throw new n8n_workflow_1.ApplicationError(`Could not connect to n8n message broker ${opts.n8nUri}: ${error.message}`, {
            cause: error,
        });
    }
}
