// @n8n/task-runner/dist/start.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const n8n_workflow_1 = require("n8n-workflow");
const a = __importStar(require("node:assert/strict"));
const authenticator_1 = require("./authenticator");
const js_task_runner_1 = require("./js-task-runner/js-task-runner");
let runner;
let isShuttingDown = false;
function readAndParseConfig() {
    const authToken = process.env.N8N_RUNNERS_AUTH_TOKEN;
    const grantToken = process.env.N8N_RUNNERS_GRANT_TOKEN;
    if (!authToken && !grantToken) {
        throw new n8n_workflow_1.ApplicationError('Missing task runner authentication. Use either N8N_RUNNERS_AUTH_TOKEN or N8N_RUNNERS_GRANT_TOKEN to configure it');
    }
    return {
        n8nUri: process.env.N8N_RUNNERS_N8N_URI ?? '127.0.0.1:5679',
        authToken,
        grantToken,
    };
}
function createSignalHandler(signal) {
    return async function onSignal() {
        if (isShuttingDown) {
            return;
        }
        console.log(`Received ${signal} signal, shutting down...`);
        isShuttingDown = true;
        try {
            if (runner) {
                await runner.stop();
                runner = undefined;
            }
        }
        catch (e) {
            const error = (0, n8n_workflow_1.ensureError)(e);
            console.error('Error stopping task runner', { error });
        }
        finally {
            process.exit(0);
        }
    };
}
void (async function start() {
    const config = readAndParseConfig();
    let grantToken = config.grantToken;
    if (!grantToken) {
        a.ok(config.authToken);
        grantToken = await (0, authenticator_1.authenticate)({
            authToken: config.authToken,
            n8nUri: config.n8nUri,
        });
    }
    const wsUrl = `ws://${config.n8nUri}/runners/_ws`;
    runner = new js_task_runner_1.JsTaskRunner({
        wsUrl,
        grantToken,
        maxConcurrency: 5,
        allowedBuiltInModules: process.env.NODE_FUNCTION_ALLOW_BUILTIN,
        allowedExternalModules: process.env.NODE_FUNCTION_ALLOW_EXTERNAL,
    });
    process.on('SIGINT', createSignalHandler('SIGINT'));
    process.on('SIGTERM', createSignalHandler('SIGTERM'));
})().catch((e) => {
    const error = (0, n8n_workflow_1.ensureError)(e);
    console.error('Task runner failed to start', { error });
    process.exit(1);
});
