// n8n/dist/execution-lifecycle-hooks/to-save-settings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSaveSettings = toSaveSettings;
const config_1 = __importDefault(require("../config"));
function toSaveSettings(workflowSettings = {}) {
    const DEFAULTS = {
        ERROR: config_1.default.getEnv('executions.saveDataOnError'),
        SUCCESS: config_1.default.getEnv('executions.saveDataOnSuccess'),
        MANUAL: config_1.default.getEnv('executions.saveDataManualExecutions'),
        PROGRESS: config_1.default.getEnv('executions.saveExecutionProgress'),
    };
    return {
        error: workflowSettings.saveDataErrorExecution
            ? workflowSettings.saveDataErrorExecution !== 'none'
            : DEFAULTS.ERROR !== 'none',
        success: workflowSettings.saveDataSuccessExecution
            ? workflowSettings.saveDataSuccessExecution !== 'none'
            : DEFAULTS.SUCCESS !== 'none',
        manual: workflowSettings === undefined || workflowSettings.saveManualExecutions === 'DEFAULT'
            ? DEFAULTS.MANUAL
            : (workflowSettings.saveManualExecutions ?? DEFAULTS.MANUAL),
        progress: workflowSettings === undefined || workflowSettings.saveExecutionProgress === 'DEFAULT'
            ? DEFAULTS.PROGRESS
            : (workflowSettings.saveExecutionProgress ?? DEFAULTS.PROGRESS),
    };
}
