// n8n-core/dist/PartialExecutionUtils/recreateNodeExecutionStack.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.recreateNodeExecutionStack = recreateNodeExecutionStack;
const a = __importStar(require("assert/strict"));
const getIncomingData_1 = require("./getIncomingData");
const getSourceDataGroups_1 = require("./getSourceDataGroups");
function recreateNodeExecutionStack(graph, startNodes, destinationNode, runData, pinData) {
    for (const node of graph.getNodes().values()) {
        a.notEqual(node.disabled, true, `Graph contains disabled nodes. This is not supported. Make sure to pass the graph through "findSubgraph" before calling "recreateNodeExecutionStack". The node in question is "${node.name}"`);
    }
    const nodeExecutionStack = [];
    const waitingExecution = {};
    const waitingExecutionSource = {};
    const runIndex = 0;
    for (const startNode of startNodes) {
        const incomingStartNodeConnections = graph
            .getDirectParentConnections(startNode)
            .filter((c) => c.type === "main");
        let incomingData = [];
        let incomingSourceData = null;
        if (incomingStartNodeConnections.length === 0) {
            incomingData.push([{ json: {} }]);
            const executeData = {
                node: startNode,
                data: { main: incomingData },
                source: incomingSourceData,
            };
            nodeExecutionStack.push(executeData);
        }
        else {
            const sourceDataSets = (0, getSourceDataGroups_1.getSourceDataGroups)(graph, startNode, runData, pinData);
            for (const sourceData of sourceDataSets) {
                incomingData = [];
                incomingSourceData = { main: [] };
                for (const incomingConnection of sourceData) {
                    const node = incomingConnection.from;
                    if (pinData[node.name]) {
                        incomingData.push(pinData[node.name]);
                    }
                    else {
                        a.ok(runData[node.name], `Start node(${incomingConnection.to.name}) has an incoming connection with no run or pinned data. This is not supported. The connection in question is "${node.name}->${startNode.name}". Are you sure the start nodes come from the "findStartNodes" function?`);
                        const nodeIncomingData = (0, getIncomingData_1.getIncomingData)(runData, node.name, runIndex, incomingConnection.type, incomingConnection.outputIndex);
                        if (nodeIncomingData) {
                            incomingData.push(nodeIncomingData);
                        }
                    }
                    incomingSourceData.main.push({
                        previousNode: incomingConnection.from.name,
                        previousNodeOutput: incomingConnection.outputIndex,
                        previousNodeRun: 0,
                    });
                }
                const executeData = {
                    node: startNode,
                    data: { main: incomingData },
                    source: incomingSourceData,
                };
                nodeExecutionStack.push(executeData);
            }
        }
        if (destinationNode) {
            const destinationNodeName = destinationNode.name;
            const incomingDestinationNodeConnections = graph
                .getDirectParentConnections(destinationNode)
                .filter((c) => c.type === "main");
            if (incomingDestinationNodeConnections !== undefined) {
                for (const connection of incomingDestinationNodeConnections) {
                    if (waitingExecution[destinationNodeName] === undefined) {
                        waitingExecution[destinationNodeName] = {};
                        waitingExecutionSource[destinationNodeName] = {};
                    }
                    if (waitingExecution[destinationNodeName][runIndex] === undefined) {
                        waitingExecution[destinationNodeName][runIndex] = {};
                        waitingExecutionSource[destinationNodeName][runIndex] = {};
                    }
                    if (waitingExecution[destinationNodeName][runIndex][connection.type] === undefined) {
                        waitingExecution[destinationNodeName][runIndex][connection.type] = [];
                        waitingExecutionSource[destinationNodeName][runIndex][connection.type] = [];
                    }
                    if (runData[connection.from.name] !== undefined) {
                        waitingExecution[destinationNodeName][runIndex][connection.type].push(runData[connection.from.name][runIndex].data[connection.type][connection.inputIndex]);
                        waitingExecutionSource[destinationNodeName][runIndex][connection.type].push({
                            previousNode: connection.from.name,
                            previousNodeOutput: connection.inputIndex || undefined,
                            previousNodeRun: runIndex || undefined,
                        });
                    }
                    else {
                        waitingExecution[destinationNodeName][runIndex][connection.type].push(null);
                        waitingExecutionSource[destinationNodeName][runIndex][connection.type].push(null);
                    }
                }
            }
        }
    }
    return {
        nodeExecutionStack,
        waitingExecution,
        waitingExecutionSource,
    };
}
