// @n8n/config/dist/configs/runners.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskRunnersConfig = void 0;
const decorators_1 = require("../decorators");
let TaskRunnersConfig = class TaskRunnersConfig {
    constructor() {
        this.disabled = true;
        this.path = '/runners';
        this.authToken = '';
        this.port = 5679;
        this.listen_address = '127.0.0.1';
        this.useLauncher = false;
        this.launcherPath = '';
        this.launcherRunner = 'javascript';
    }
};
exports.TaskRunnersConfig = TaskRunnersConfig;
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_DISABLED'),
    __metadata("design:type", Boolean)
], TaskRunnersConfig.prototype, "disabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_PATH'),
    __metadata("design:type", String)
], TaskRunnersConfig.prototype, "path", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_AUTH_TOKEN'),
    __metadata("design:type", String)
], TaskRunnersConfig.prototype, "authToken", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_SERVER_PORT'),
    __metadata("design:type", Number)
], TaskRunnersConfig.prototype, "port", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_SERVER_LISTEN_ADDRESS'),
    __metadata("design:type", String)
], TaskRunnersConfig.prototype, "listen_address", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_USE_LAUNCHER'),
    __metadata("design:type", Boolean)
], TaskRunnersConfig.prototype, "useLauncher", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_LAUNCHER_PATH'),
    __metadata("design:type", String)
], TaskRunnersConfig.prototype, "launcherPath", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_RUNNERS_LAUNCHER_RUNNER'),
    __metadata("design:type", String)
], TaskRunnersConfig.prototype, "launcherRunner", void 0);
exports.TaskRunnersConfig = TaskRunnersConfig = __decorate([
    decorators_1.Config
], TaskRunnersConfig);
