// @n8n/task-runner/dist/js-task-runner/js-task-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsTaskRunner = void 0;
const n8n_core_1 = require("n8n-core");
const n8n_workflow_1 = require("n8n-workflow");
const a = __importStar(require("node:assert"));
const node_vm_1 = require("node:vm");
const task_runner_1 = require("../task-runner");
const error_like_1 = require("./errors/error-like");
const execution_error_1 = require("./errors/execution-error");
const require_resolver_1 = require("./require-resolver");
const result_validation_1 = require("./result-validation");
class JsTaskRunner extends task_runner_1.TaskRunner {
    constructor({ grantToken, maxConcurrency, wsUrl, name = 'JS Task Runner', allowedBuiltInModules, allowedExternalModules, }) {
        super('javascript', wsUrl, grantToken, maxConcurrency, name);
        const parseModuleAllowList = (moduleList) => moduleList === '*' ? null : new Set(moduleList.split(',').map((x) => x.trim()));
        this.requireResolver = (0, require_resolver_1.createRequireResolver)({
            allowedBuiltInModules: parseModuleAllowList(allowedBuiltInModules ?? ''),
            allowedExternalModules: parseModuleAllowList(allowedExternalModules ?? ''),
        });
    }
    async executeTask(task) {
        const allData = await this.requestData(task.taskId, 'all');
        const settings = task.settings;
        a.ok(settings, 'JS Code not sent to runner');
        const workflowParams = allData.workflow;
        const workflow = new n8n_workflow_1.Workflow({
            ...workflowParams,
            nodeTypes: {
                getByNameAndVersion() {
                    return undefined;
                },
                getByName() {
                    return undefined;
                },
                getKnownTypes() {
                    return {};
                },
            },
        });
        const customConsole = {
            log: (...args) => {
                const logOutput = args
                    .map((arg) => (typeof arg === 'object' && arg !== null ? JSON.stringify(arg) : arg))
                    .join(' ');
                void this.makeRpcCall(task.taskId, 'logNodeOutput', [logOutput]);
            },
        };
        const result = settings.nodeMode === 'runOnceForAllItems'
            ? await this.runForAllItems(task.taskId, settings, allData, workflow, customConsole)
            : await this.runForEachItem(task.taskId, settings, allData, workflow, customConsole);
        return {
            result,
            customData: allData.runExecutionData.resultData.metadata,
        };
    }
    async runForAllItems(taskId, settings, allData, workflow, customConsole) {
        const dataProxy = this.createDataProxy(allData, workflow, allData.itemIndex);
        const inputItems = allData.connectionInputData;
        const context = {
            require: this.requireResolver,
            module: {},
            console: customConsole,
            items: inputItems,
            ...dataProxy,
            ...this.buildRpcCallObject(taskId),
        };
        try {
            const result = (await (0, node_vm_1.runInNewContext)(`module.exports = async function VmCodeWrapper() {${settings.code}\n}()`, context));
            if (result === null) {
                return [];
            }
            return (0, result_validation_1.validateRunForAllItemsOutput)(result);
        }
        catch (e) {
            const error = this.toExecutionErrorIfNeeded(e);
            if (settings.continueOnFail) {
                return [{ json: { error: error.message } }];
            }
            throw error;
        }
    }
    async runForEachItem(taskId, settings, allData, workflow, customConsole) {
        const inputItems = allData.connectionInputData;
        const returnData = [];
        for (let index = 0; index < inputItems.length; index++) {
            const item = inputItems[index];
            const dataProxy = this.createDataProxy(allData, workflow, index);
            const context = {
                require: this.requireResolver,
                module: {},
                console: customConsole,
                item,
                ...dataProxy,
                ...this.buildRpcCallObject(taskId),
            };
            try {
                let result = (await (0, node_vm_1.runInNewContext)(`module.exports = async function VmCodeWrapper() {${settings.code}\n}()`, context));
                if (result === null) {
                    continue;
                }
                result = (0, result_validation_1.validateRunForEachItemOutput)(result, index);
                if (result) {
                    returnData.push(result.binary
                        ? {
                            json: result.json,
                            pairedItem: { item: index },
                            binary: result.binary,
                        }
                        : {
                            json: result.json,
                            pairedItem: { item: index },
                        });
                }
            }
            catch (e) {
                const error = this.toExecutionErrorIfNeeded(e);
                if (!settings.continueOnFail) {
                    throw error;
                }
                returnData.push({
                    json: { error: error.message },
                    pairedItem: {
                        item: index,
                    },
                });
            }
        }
        return returnData;
    }
    createDataProxy(allData, workflow, itemIndex) {
        return new n8n_workflow_1.WorkflowDataProxy(workflow, allData.runExecutionData, allData.runIndex, itemIndex, allData.activeNodeName, allData.connectionInputData, allData.siblingParameters, allData.mode, (0, n8n_core_1.getAdditionalKeys)(allData.additionalData, allData.mode, allData.runExecutionData), allData.executeData, allData.defaultReturnRunIndex, allData.selfData, allData.contextNodeName, allData.envProviderState ?? {
            env: {},
            isEnvAccessBlocked: false,
            isProcessAvailable: true,
        }).getDataProxy();
    }
    toExecutionErrorIfNeeded(error) {
        if (error instanceof Error) {
            return error;
        }
        if ((0, error_like_1.isErrorLike)(error)) {
            return new execution_error_1.ExecutionError(error);
        }
        return new execution_error_1.ExecutionError({ message: JSON.stringify(error) });
    }
}
exports.JsTaskRunner = JsTaskRunner;
