// n8n/dist/services/public-api-key.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicApiKeyService = exports.API_KEY_PREFIX = void 0;
const node_crypto_1 = require("node:crypto");
const typedi_1 = __importStar(require("typedi"));
const api_key_1 = require("../databases/entities/api-key");
const api_key_repository_1 = require("../databases/repositories/api-key.repository");
const user_repository_1 = require("../databases/repositories/user.repository");
exports.API_KEY_PREFIX = 'n8n_api_';
let PublicApiKeyService = class PublicApiKeyService {
    constructor(apiKeyRepository) {
        this.apiKeyRepository = apiKeyRepository;
        this.createApiKeyString = () => `${exports.API_KEY_PREFIX}${(0, node_crypto_1.randomBytes)(40).toString('hex')}`;
    }
    async createPublicApiKeyForUser(user) {
        const apiKey = this.createApiKeyString();
        await this.apiKeyRepository.upsert(this.apiKeyRepository.create({
            userId: user.id,
            apiKey,
            label: 'My API Key',
        }), ['apiKey']);
        return await this.apiKeyRepository.findOneByOrFail({ apiKey });
    }
    async getRedactedApiKeysForUser(user) {
        const apiKeys = await this.apiKeyRepository.findBy({ userId: user.id });
        return apiKeys.map((apiKeyRecord) => ({
            ...apiKeyRecord,
            apiKey: this.redactApiKey(apiKeyRecord.apiKey),
        }));
    }
    async deleteApiKeyForUser(user, apiKeyId) {
        await this.apiKeyRepository.delete({ userId: user.id, id: apiKeyId });
    }
    async getUserForApiKey(apiKey) {
        return await typedi_1.default.get(user_repository_1.UserRepository)
            .createQueryBuilder('user')
            .innerJoin(api_key_1.ApiKey, 'apiKey', 'apiKey.userId = user.id')
            .where('apiKey.apiKey = :apiKey', { apiKey })
            .select('user')
            .getOne();
    }
    redactApiKey(apiKey) {
        const keepLength = 5;
        return (exports.API_KEY_PREFIX +
            apiKey.slice(exports.API_KEY_PREFIX.length, exports.API_KEY_PREFIX.length + keepLength) +
            '*'.repeat(apiKey.length - exports.API_KEY_PREFIX.length - keepLength));
    }
};
exports.PublicApiKeyService = PublicApiKeyService;
exports.PublicApiKeyService = PublicApiKeyService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [api_key_repository_1.ApiKeyRepository])
], PublicApiKeyService);
