// n8n-editor-ui/dist/assets/FixedCollectionParameter-h3uHE3EI.js
import { d as defineComponent, dK as isINodePropertyCollectionList, dF as deepCopy, dD as get, _ as _export_sfc, l as resolveComponent, c as openBlock, h as createElementBlock, i as createVNode, w as withCtx, k as createTextVNode, t as toDisplayString, f as createCommentVNode, F as Fragment, z as renderList, e as createBlock, j as createBaseVNode, aS as Suspense, n as normalizeClass, I as withModifiers, dz as _sfc_main$1 } from "./index-TQ22MZub.js";
const _sfc_main = defineComponent({
  name: "FixedCollectionParameter",
  props: {
    nodeValues: {
      type: Object,
      required: true
    },
    parameter: {
      type: Object,
      required: true
    },
    path: {
      type: String,
      required: true
    },
    values: {
      type: Object,
      default: () => ({})
    },
    isReadOnly: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      selectedOption: void 0,
      mutableValues: {}
    };
  },
  computed: {
    getPlaceholderText() {
      const placeholder = this.$locale.nodeText().placeholder(this.parameter, this.path);
      return placeholder ? placeholder : this.$locale.baseText("fixedCollectionParameter.choose");
    },
    getProperties() {
      const returnProperties = [];
      let tempProperties;
      for (const name of this.propertyNames) {
        tempProperties = this.getOptionProperties(name);
        if (tempProperties !== void 0) {
          returnProperties.push(tempProperties);
        }
      }
      return returnProperties;
    },
    multipleValues() {
      var _a;
      return !!((_a = this.parameter.typeOptions) == null ? void 0 : _a.multipleValues);
    },
    parameterOptions() {
      if (this.multipleValues && isINodePropertyCollectionList(this.parameter.options)) {
        return this.parameter.options;
      }
      return this.parameter.options.filter((option) => {
        return !this.propertyNames.includes(option.name);
      });
    },
    propertyNames() {
      return Object.keys(this.mutableValues || {});
    },
    sortable() {
      var _a;
      return !!((_a = this.parameter.typeOptions) == null ? void 0 : _a.sortable);
    }
  },
  watch: {
    values: {
      handler(newValues) {
        this.mutableValues = deepCopy(newValues);
      },
      deep: true
    }
  },
  created() {
    this.mutableValues = deepCopy(this.values);
  },
  methods: {
    deleteOption(optionName, index) {
      const currentOptionsOfSameType = this.mutableValues[optionName];
      if (!currentOptionsOfSameType || currentOptionsOfSameType.length > 1) {
        this.$emit("valueChanged", {
          name: this.getPropertyPath(optionName, index),
          value: void 0
        });
      } else {
        this.$emit("valueChanged", {
          name: this.getPropertyPath(optionName),
          value: void 0
        });
      }
    },
    getPropertyPath(name, index) {
      return `${this.path}.${name}` + (index !== void 0 ? `[${index}]` : "");
    },
    getOptionProperties(optionName) {
      if (isINodePropertyCollectionList(this.parameter.options)) {
        for (const option of this.parameter.options) {
          if (option.name === optionName) {
            return option;
          }
        }
      }
      return void 0;
    },
    moveOptionDown(optionName, index) {
      if (Array.isArray(this.mutableValues[optionName])) {
        this.mutableValues[optionName].splice(
          index + 1,
          0,
          this.mutableValues[optionName].splice(index, 1)[0]
        );
      }
      const parameterData = {
        name: this.getPropertyPath(optionName),
        value: this.mutableValues[optionName],
        type: "optionsOrderChanged"
      };
      this.$emit("valueChanged", parameterData);
    },
    moveOptionUp(optionName, index) {
      var _a;
      if (Array.isArray(this.mutableValues[optionName])) {
        (_a = this.mutableValues) == null ? void 0 : _a[optionName].splice(
          index - 1,
          0,
          this.mutableValues[optionName].splice(index, 1)[0]
        );
      }
      const parameterData = {
        name: this.getPropertyPath(optionName),
        value: this.mutableValues[optionName],
        type: "optionsOrderChanged"
      };
      this.$emit("valueChanged", parameterData);
    },
    optionSelected(optionName) {
      const option = this.getOptionProperties(optionName);
      if (option === void 0) {
        return;
      }
      const name = `${this.path}.${option.name}`;
      const newParameterValue = {};
      for (const optionParameter of option.values) {
        if (optionParameter.type === "fixedCollection" && optionParameter.typeOptions !== void 0 && optionParameter.typeOptions.multipleValues === true) {
          newParameterValue[optionParameter.name] = {};
        } else if (optionParameter.typeOptions !== void 0 && optionParameter.typeOptions.multipleValues === true) {
          const multiValue = get(this.nodeValues, [this.path, optionParameter.name], []);
          if (Array.isArray(optionParameter.default)) {
            multiValue.push(...deepCopy(optionParameter.default));
          } else if (optionParameter.default !== "" && typeof optionParameter.default !== "object") {
            multiValue.push(deepCopy(optionParameter.default));
          }
          newParameterValue[optionParameter.name] = multiValue;
        } else {
          newParameterValue[optionParameter.name] = deepCopy(optionParameter.default);
        }
      }
      let newValue;
      if (this.multipleValues) {
        newValue = get(this.nodeValues, name, []);
        newValue.push(newParameterValue);
      } else {
        newValue = newParameterValue;
      }
      const parameterData = {
        name,
        value: newValue
      };
      this.$emit("valueChanged", parameterData);
      this.selectedOption = void 0;
    },
    valueChanged(parameterData) {
      this.$emit("valueChanged", parameterData);
    }
  }
});
const _hoisted_1 = ["data-test-id"];
const _hoisted_2 = {
  key: 0,
  class: "no-items-exist"
};
const _hoisted_3 = { key: 1 };
const _hoisted_4 = {
  key: 0,
  class: "delete-option"
};
const _hoisted_5 = {
  key: 2,
  class: "parameter-item"
};
const _hoisted_6 = { class: "parameter-item-wrapper" };
const _hoisted_7 = {
  key: 0,
  class: "delete-option"
};
const _hoisted_8 = {
  key: 1,
  class: "controls"
};
const _hoisted_9 = {
  key: 1,
  class: "add-option"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_n8n_input_label = resolveComponent("n8n-input-label");
  const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
  const _component_ParameterInputList = _sfc_main$1;
  const _component_n8n_button = resolveComponent("n8n-button");
  const _component_n8n_option = resolveComponent("n8n-option");
  const _component_n8n_select = resolveComponent("n8n-select");
  return openBlock(), createElementBlock("div", {
    class: "fixed-collection-parameter",
    "data-test-id": `fixed-collection-${_ctx.parameter.name}`,
    onKeydown: _cache[2] || (_cache[2] = withModifiers(() => {
    }, ["stop"]))
  }, [
    _ctx.getProperties.length === 0 ? (openBlock(), createElementBlock("div", _hoisted_2, [
      createVNode(_component_n8n_text, { size: "small" }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("fixedCollectionParameter.currentlyNoItemsExist")), 1)
        ]),
        _: 1
      })
    ])) : createCommentVNode("", true),
    (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.getProperties, (property) => {
      return openBlock(), createElementBlock("div", {
        key: property.name,
        class: "fixed-collection-parameter-property"
      }, [
        property.displayName !== "" && _ctx.parameter.options && _ctx.parameter.options.length !== 1 ? (openBlock(), createBlock(_component_n8n_input_label, {
          key: 0,
          label: _ctx.$locale.nodeText().inputLabelDisplayName(property, _ctx.path),
          underline: true,
          size: "small",
          color: "text-dark"
        }, null, 8, ["label"])) : createCommentVNode("", true),
        _ctx.multipleValues ? (openBlock(), createElementBlock("div", _hoisted_3, [
          (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.mutableValues[property.name], (_, index) => {
            return openBlock(), createElementBlock("div", {
              key: property.name + index,
              class: "parameter-item"
            }, [
              createBaseVNode("div", {
                class: normalizeClass(index ? "border-top-dashed parameter-item-wrapper " : "parameter-item-wrapper")
              }, [
                !_ctx.isReadOnly ? (openBlock(), createElementBlock("div", _hoisted_4, [
                  createVNode(_component_n8n_icon_button, {
                    type: "tertiary",
                    text: "",
                    size: "mini",
                    icon: "trash",
                    title: _ctx.$locale.baseText("fixedCollectionParameter.deleteItem"),
                    onClick: ($event) => _ctx.deleteOption(property.name, index)
                  }, null, 8, ["title", "onClick"]),
                  _ctx.sortable && index !== 0 ? (openBlock(), createBlock(_component_n8n_icon_button, {
                    key: 0,
                    type: "tertiary",
                    text: "",
                    size: "mini",
                    icon: "angle-up",
                    title: _ctx.$locale.baseText("fixedCollectionParameter.moveUp"),
                    onClick: ($event) => _ctx.moveOptionUp(property.name, index)
                  }, null, 8, ["title", "onClick"])) : createCommentVNode("", true),
                  _ctx.sortable && index !== _ctx.mutableValues[property.name].length - 1 ? (openBlock(), createBlock(_component_n8n_icon_button, {
                    key: 1,
                    type: "tertiary",
                    text: "",
                    size: "mini",
                    icon: "angle-down",
                    title: _ctx.$locale.baseText("fixedCollectionParameter.moveDown"),
                    onClick: ($event) => _ctx.moveOptionDown(property.name, index)
                  }, null, 8, ["title", "onClick"])) : createCommentVNode("", true)
                ])) : createCommentVNode("", true),
                (openBlock(), createBlock(Suspense, null, {
                  default: withCtx(() => [
                    createVNode(_component_ParameterInputList, {
                      parameters: property.values,
                      "node-values": _ctx.nodeValues,
                      path: _ctx.getPropertyPath(property.name, index),
                      "hide-delete": true,
                      "is-read-only": _ctx.isReadOnly,
                      onValueChanged: _ctx.valueChanged
                    }, null, 8, ["parameters", "node-values", "path", "is-read-only", "onValueChanged"])
                  ]),
                  _: 2
                }, 1024))
              ], 2)
            ]);
          }), 128))
        ])) : (openBlock(), createElementBlock("div", _hoisted_5, [
          createBaseVNode("div", _hoisted_6, [
            !_ctx.isReadOnly ? (openBlock(), createElementBlock("div", _hoisted_7, [
              createVNode(_component_n8n_icon_button, {
                type: "tertiary",
                text: "",
                size: "mini",
                icon: "trash",
                title: _ctx.$locale.baseText("fixedCollectionParameter.deleteItem"),
                onClick: ($event) => _ctx.deleteOption(property.name)
              }, null, 8, ["title", "onClick"])
            ])) : createCommentVNode("", true),
            createVNode(_component_ParameterInputList, {
              parameters: property.values,
              "node-values": _ctx.nodeValues,
              path: _ctx.getPropertyPath(property.name),
              "is-read-only": _ctx.isReadOnly,
              class: "parameter-item",
              "hide-delete": true,
              onValueChanged: _ctx.valueChanged
            }, null, 8, ["parameters", "node-values", "path", "is-read-only", "onValueChanged"])
          ])
        ]))
      ]);
    }), 128)),
    _ctx.parameterOptions.length > 0 && !_ctx.isReadOnly ? (openBlock(), createElementBlock("div", _hoisted_8, [
      _ctx.parameter.options && _ctx.parameter.options.length === 1 ? (openBlock(), createBlock(_component_n8n_button, {
        key: 0,
        type: "tertiary",
        block: "",
        "data-test-id": "fixed-collection-add",
        label: _ctx.getPlaceholderText,
        onClick: _cache[0] || (_cache[0] = ($event) => _ctx.optionSelected(_ctx.parameter.options[0].name))
      }, null, 8, ["label"])) : (openBlock(), createElementBlock("div", _hoisted_9, [
        createVNode(_component_n8n_select, {
          modelValue: _ctx.selectedOption,
          "onUpdate:modelValue": [
            _cache[1] || (_cache[1] = ($event) => _ctx.selectedOption = $event),
            _ctx.optionSelected
          ],
          placeholder: _ctx.getPlaceholderText,
          size: "small",
          filterable: ""
        }, {
          default: withCtx(() => [
            (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.parameterOptions, (item) => {
              return openBlock(), createBlock(_component_n8n_option, {
                key: item.name,
                label: _ctx.$locale.nodeText().collectionOptionDisplayName(_ctx.parameter, item, _ctx.path),
                value: item.name
              }, null, 8, ["label", "value"]);
            }), 128))
          ]),
          _: 1
        }, 8, ["modelValue", "placeholder", "onUpdate:modelValue"])
      ]))
    ])) : createCommentVNode("", true)
  ], 40, _hoisted_1);
}
const FixedCollectionParameter = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__scopeId", "data-v-f59887d3"]]);
export {
  FixedCollectionParameter as default
};
