// n8n-editor-ui/dist/assets/ProjectCardBadge.vue_vue_type_script_setup_true_lang-R4VvIcND.js
import { d as defineComponent, p as computed, dA as ProjectTypes, iS as splitName, l as resolveComponent, c9 as resolveDirective, c as openBlock, e as createBlock, w as withCtx, k as createTextVNode, t as toDisplayString, as as withDirectives, h as createElementBlock, f as createCommentVNode, g as useI18n } from "./index-TQ22MZub.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "ProjectCardBadge",
  props: {
    resource: {},
    resourceType: {},
    resourceTypeLabel: {},
    personalProject: {}
  },
  setup(__props) {
    const props = __props;
    const i18n = useI18n();
    const projectState = computed(() => {
      var _a, _b, _c, _d, _e;
      if (props.resource.homeProject && props.personalProject && props.resource.homeProject.id === props.personalProject.id || !props.resource.homeProject) {
        if ((_a = props.resource.sharedWithProjects) == null ? void 0 : _a.length) {
          return "shared-owned";
        }
        return "owned";
      } else if (((_b = props.resource.homeProject) == null ? void 0 : _b.type) !== ProjectTypes.Team) {
        if ((_c = props.resource.sharedWithProjects) == null ? void 0 : _c.length) {
          return "shared-personal";
        }
        return "personal";
      } else if (((_d = props.resource.homeProject) == null ? void 0 : _d.type) === ProjectTypes.Team) {
        if ((_e = props.resource.sharedWithProjects) == null ? void 0 : _e.length) {
          return "shared-team";
        }
        return "team";
      }
      return "unknown";
    });
    const badgeText = computed(() => {
      var _a;
      if (projectState.value === "owned" || projectState.value === "shared-owned") {
        return i18n.baseText("generic.ownedByMe");
      } else {
        const { name, email } = splitName(((_a = props.resource.homeProject) == null ? void 0 : _a.name) ?? "");
        return name ?? email ?? "";
      }
    });
    const badgeIcon = computed(() => {
      switch (projectState.value) {
        case "shared-personal":
        case "shared-owned":
          return "user-friends";
        case "team":
        case "shared-team":
          return "archive";
        default:
          return "";
      }
    });
    const badgeTooltip = computed(() => {
      switch (projectState.value) {
        case "shared-owned":
          return i18n.baseText("projects.badge.tooltip.sharedOwned", {
            interpolate: {
              resourceTypeLabel: props.resourceTypeLabel
            }
          });
        case "shared-personal":
          return i18n.baseText("projects.badge.tooltip.sharedPersonal", {
            interpolate: {
              resourceTypeLabel: props.resourceTypeLabel,
              name: badgeText.value
            }
          });
        case "personal":
          return i18n.baseText("projects.badge.tooltip.personal", {
            interpolate: {
              resourceTypeLabel: props.resourceTypeLabel,
              name: badgeText.value
            }
          });
        case "team":
          return i18n.baseText("projects.badge.tooltip.team", {
            interpolate: {
              resourceTypeLabel: props.resourceTypeLabel,
              name: badgeText.value
            }
          });
        case "shared-team":
          return i18n.baseText("projects.badge.tooltip.sharedTeam", {
            interpolate: {
              resourceTypeLabel: props.resourceTypeLabel,
              name: badgeText.value
            }
          });
        default:
          return "";
      }
    });
    return (_ctx, _cache) => {
      const _component_N8nIcon = resolveComponent("N8nIcon");
      const _component_N8nBadge = resolveComponent("N8nBadge");
      const _component_N8nTooltip = resolveComponent("N8nTooltip");
      const _directive_n8n_truncate = resolveDirective("n8n-truncate");
      return openBlock(), createBlock(_component_N8nTooltip, {
        disabled: !badgeTooltip.value,
        placement: "top"
      }, {
        content: withCtx(() => [
          createTextVNode(toDisplayString(badgeTooltip.value), 1)
        ]),
        default: withCtx(() => [
          badgeText.value ? (openBlock(), createBlock(_component_N8nBadge, {
            key: 0,
            class: "mr-xs",
            theme: "tertiary",
            bold: "",
            "data-test-id": "card-badge"
          }, {
            default: withCtx(() => [
              withDirectives((openBlock(), createElementBlock("span", null, [
                createTextVNode(toDisplayString(badgeText.value), 1)
              ])), [
                [_directive_n8n_truncate, void 0, "20"]
              ]),
              badgeIcon.value ? (openBlock(), createBlock(_component_N8nIcon, {
                key: 0,
                icon: badgeIcon.value,
                size: "small",
                class: "ml-5xs"
              }, null, 8, ["icon"])) : createCommentVNode("", true)
            ]),
            _: 1
          })) : createCommentVNode("", true)
        ]),
        _: 1
      }, 8, ["disabled"]);
    };
  }
});
export {
  _sfc_main as _
};
