// n8n-editor-ui/dist/assets/ResourcesListLayout-Cp1vFuQD.js
import { d as defineComponent, e0 as mapStores, K as useUIStore, _ as _export_sfc, c as openBlock, h as createElementBlock, q as renderSlot, j as createBaseVNode, n as normalizeClass, f as createCommentVNode, ix as ProjectSharing, a8 as EnterpriseEditionFeature, a1 as useProjectsStore, e as createBlock, w as withCtx, i as createVNode, as as withDirectives, at as vShow, k as createTextVNode, t as toDisplayString, l as resolveComponent, U as useRoute, J as useDebounce, u as useUsersStore, r as ref, p as computed, H as watch, o as onMounted, C as nextTick, g as useI18n, ak as useTelemetry, bQ as normalizeProps, bs as guardReactiveProps, F as Fragment, z as renderList } from "./index-TQ22MZub.js";
const _sfc_main$3 = defineComponent({
  name: "PageViewLayout",
  data() {
    return {
      loading: false
    };
  },
  computed: {
    ...mapStores(useUIStore)
  }
});
const wrapper$1 = "_wrapper_u24kg_1";
const content = "_content_u24kg_12";
const style0$3 = {
  wrapper: wrapper$1,
  content
};
function _sfc_render$3(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.$style.wrapper)
  }, [
    renderSlot(_ctx.$slots, "header"),
    createBaseVNode("main", {
      class: normalizeClass(_ctx.$style.content)
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 2)
  ], 2);
}
const cssModules$3 = {
  "$style": style0$3
};
const __unplugin_components_2 = /* @__PURE__ */ _export_sfc(_sfc_main$3, [["render", _sfc_render$3], ["__cssModules", cssModules$3]]);
const _sfc_main$2 = defineComponent({
  props: {
    overflow: {
      type: Boolean,
      default: false
    }
  }
});
const wrapper = "_wrapper_183my_1";
const overflow = "_overflow_183my_6";
const list = "_list_183my_6";
const body = "_body_183my_6";
const style0$2 = {
  wrapper,
  overflow,
  list,
  body
};
function _sfc_render$2(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass({ [_ctx.$style.wrapper]: true, [_ctx.$style.overflow]: _ctx.overflow })
  }, [
    createBaseVNode("div", {
      class: normalizeClass(_ctx.$style.list)
    }, [
      _ctx.$slots.header ? (openBlock(), createElementBlock("div", {
        key: 0,
        class: normalizeClass(_ctx.$style.header)
      }, [
        renderSlot(_ctx.$slots, "header")
      ], 2)) : createCommentVNode("", true),
      createBaseVNode("div", {
        class: normalizeClass(_ctx.$style.body)
      }, [
        renderSlot(_ctx.$slots, "default")
      ], 2)
    ], 2)
  ], 2);
}
const cssModules$2 = {
  "$style": style0$2
};
const __unplugin_components_1 = /* @__PURE__ */ _export_sfc(_sfc_main$2, [["render", _sfc_render$2], ["__cssModules", cssModules$2]]);
const _sfc_main$1 = defineComponent({
  components: {
    ProjectSharing
  },
  props: {
    modelValue: {
      type: Object,
      default: () => ({})
    },
    keys: {
      type: Array,
      default: () => []
    },
    shareable: {
      type: Boolean,
      default: true
    },
    reset: {
      type: Function,
      default: () => {
      }
    }
  },
  data() {
    return {
      EnterpriseEditionFeature,
      selectedProject: null
    };
  },
  computed: {
    ...mapStores(useProjectsStore),
    filtersLength() {
      let length = 0;
      this.keys.forEach((key) => {
        if (key === "search") {
          return;
        }
        const value = this.modelValue[key];
        length += (Array.isArray(value) ? value.length > 0 : value !== "") ? 1 : 0;
      });
      return length;
    },
    hasFilters() {
      return this.filtersLength > 0;
    }
  },
  watch: {
    filtersLength(value) {
      this.$emit("update:filtersLength", value);
    }
  },
  async beforeMount() {
    await this.projectsStore.getAvailableProjects();
    this.selectedProject = this.projectsStore.availableProjects.find(
      (project) => project.id === this.modelValue.homeProject
    ) ?? null;
  },
  methods: {
    setKeyValue(key, value) {
      const filters2 = {
        ...this.modelValue,
        [key]: value
      };
      this.$emit("update:modelValue", filters2);
    },
    resetFilters() {
      if (this.reset) {
        this.reset();
      } else {
        const filters2 = { ...this.modelValue };
        this.keys.forEach((key) => {
          filters2[key] = Array.isArray(this.modelValue[key]) ? [] : "";
        });
        this.$emit("update:modelValue", filters2);
      }
      this.selectedProject = null;
    }
  }
});
const style0$1 = {
  "filter-button": "_filter-button_i7gly_1",
  "filters-dropdown": "_filters-dropdown_i7gly_6",
  "filters-dropdown-footer": "_filters-dropdown-footer_i7gly_10"
};
function _sfc_render$1(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_badge = resolveComponent("n8n-badge");
  const _component_n8n_button = resolveComponent("n8n-button");
  const _component_n8n_input_label = resolveComponent("n8n-input-label");
  const _component_ProjectSharing = ProjectSharing;
  const _component_enterprise_edition = resolveComponent("enterprise-edition");
  const _component_n8n_link = resolveComponent("n8n-link");
  const _component_n8n_popover = resolveComponent("n8n-popover");
  return openBlock(), createBlock(_component_n8n_popover, {
    trigger: "click",
    width: "304",
    size: "large"
  }, {
    reference: withCtx(() => [
      createVNode(_component_n8n_button, {
        icon: "filter",
        type: "tertiary",
        active: _ctx.hasFilters,
        class: normalizeClass(_ctx.$style["filter-button"]),
        "data-test-id": "resources-list-filters-trigger"
      }, {
        default: withCtx(() => [
          withDirectives(createVNode(_component_n8n_badge, {
            theme: "primary",
            class: "mr-4xs"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.filtersLength), 1)
            ]),
            _: 1
          }, 512), [
            [vShow, _ctx.filtersLength > 0]
          ]),
          createTextVNode(" " + toDisplayString(_ctx.$locale.baseText("forms.resourceFiltersDropdown.filters")), 1)
        ]),
        _: 1
      }, 8, ["active", "class"])
    ]),
    default: withCtx(() => [
      createBaseVNode("div", {
        class: normalizeClass(_ctx.$style["filters-dropdown"]),
        "data-test-id": "resources-list-filters-dropdown"
      }, [
        renderSlot(_ctx.$slots, "default", {
          filters: _ctx.modelValue,
          setKeyValue: _ctx.setKeyValue
        }),
        _ctx.shareable && _ctx.projectsStore.isProjectHome ? (openBlock(), createBlock(_component_enterprise_edition, {
          key: 0,
          features: [_ctx.EnterpriseEditionFeature.Sharing]
        }, {
          default: withCtx(() => [
            createVNode(_component_n8n_input_label, {
              label: _ctx.$locale.baseText("forms.resourceFiltersDropdown.owner"),
              bold: false,
              size: "small",
              color: "text-base",
              class: "mb-3xs"
            }, null, 8, ["label"]),
            createVNode(_component_ProjectSharing, {
              modelValue: _ctx.selectedProject,
              "onUpdate:modelValue": [
                _cache[0] || (_cache[0] = ($event) => _ctx.selectedProject = $event),
                _cache[1] || (_cache[1] = ($event) => _ctx.setKeyValue("homeProject", $event.id))
              ],
              projects: _ctx.projectsStore.availableProjects,
              placeholder: _ctx.$locale.baseText("forms.resourceFiltersDropdown.owner.placeholder"),
              "empty-options-text": _ctx.$locale.baseText("projects.sharing.noMatchingProjects")
            }, null, 8, ["modelValue", "projects", "placeholder", "empty-options-text"])
          ]),
          _: 1
        }, 8, ["features"])) : createCommentVNode("", true),
        _ctx.hasFilters ? (openBlock(), createElementBlock("div", {
          key: 1,
          class: normalizeClass([_ctx.$style["filters-dropdown-footer"], "mt-s"])
        }, [
          createVNode(_component_n8n_link, { onClick: _ctx.resetFilters }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("forms.resourceFiltersDropdown.reset")), 1)
            ]),
            _: 1
          }, 8, ["onClick"])
        ], 2)) : createCommentVNode("", true)
      ], 2)
    ]),
    _: 3
  });
}
const cssModules$1 = {
  "$style": style0$1
};
const __unplugin_components_0 = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["render", _sfc_render$1], ["__cssModules", cssModules$1]]);
const _sfc_main = defineComponent({
  name: "ResourcesListLayout",
  components: {
    PageViewLayout: __unplugin_components_2,
    PageViewLayoutList: __unplugin_components_1,
    ResourceFiltersDropdown: __unplugin_components_0
  },
  props: {
    resourceKey: {
      type: String,
      default: ""
    },
    displayName: {
      type: Function,
      default: (resource) => resource.name
    },
    resources: {
      type: Array,
      default: () => []
    },
    disabled: {
      type: Boolean,
      default: false
    },
    initialize: {
      type: Function,
      default: () => async () => {
      }
    },
    filters: {
      type: Object,
      default: () => ({ search: "", homeProject: "" })
    },
    additionalFiltersHandler: {
      type: Function,
      required: false,
      default: void 0
    },
    shareable: {
      type: Boolean,
      default: true
    },
    showFiltersDropdown: {
      type: Boolean,
      default: true
    },
    sortFns: {
      type: Object,
      default: () => ({})
    },
    sortOptions: {
      type: Array,
      default: () => ["lastUpdated", "lastCreated", "nameAsc", "nameDesc"]
    },
    type: {
      type: String,
      default: "list"
    },
    typeProps: {
      type: Object,
      default: () => ({
        itemSize: 80
      })
    },
    loading: {
      type: Boolean,
      required: false,
      default: true
    }
  },
  emits: ["update:filters", "click:add", "sort"],
  setup(props, { emit }) {
    const route = useRoute();
    const i18n = useI18n();
    const { callDebounced } = useDebounce();
    const usersStore = useUsersStore();
    const telemetry = useTelemetry();
    const sortBy = ref(props.sortOptions[0]);
    const hasFilters = ref(false);
    const filtersModel = ref(props.filters);
    const currentPage = ref(1);
    const rowsPerPage = ref(10);
    const resettingFilters = ref(false);
    const search2 = ref(null);
    const filterKeys = computed(() => {
      return Object.keys(filtersModel.value);
    });
    const filteredAndSortedResources = computed(() => {
      const filtered = props.resources.filter((resource) => {
        let matches = true;
        if (filtersModel.value.homeProject) {
          matches = matches && !!(resource.homeProject && resource.homeProject.id === filtersModel.value.homeProject);
        }
        if (filtersModel.value.search) {
          const searchString = filtersModel.value.search.toLowerCase();
          matches = matches && props.displayName(resource).toLowerCase().includes(searchString);
        }
        if (props.additionalFiltersHandler) {
          matches = props.additionalFiltersHandler(resource, filtersModel.value, matches);
        }
        return matches;
      });
      return filtered.sort((a, b) => {
        switch (sortBy.value) {
          case "lastUpdated":
            return props.sortFns.lastUpdated ? props.sortFns.lastUpdated(a, b) : new Date(b.updatedAt ?? "").valueOf() - new Date(a.updatedAt ?? "").valueOf();
          case "lastCreated":
            return props.sortFns.lastCreated ? props.sortFns.lastCreated(a, b) : new Date(b.createdAt ?? "").valueOf() - new Date(a.createdAt ?? "").valueOf();
          case "nameAsc":
            return props.sortFns.nameAsc ? props.sortFns.nameAsc(a, b) : props.displayName(a).trim().localeCompare(props.displayName(b).trim());
          case "nameDesc":
            return props.sortFns.nameDesc ? props.sortFns.nameDesc(a, b) : props.displayName(b).trim().localeCompare(props.displayName(a).trim());
          default:
            return props.sortFns[sortBy.value] ? props.sortFns[sortBy.value](a, b) : 0;
        }
      });
    });
    const focusSearchInput = () => {
      if (search2.value) {
        search2.value.focus();
      }
    };
    const hasAppliedFilters = () => {
      return !!filterKeys.value.find(
        (key) => key !== "search" && (Array.isArray(props.filters[key]) ? props.filters[key].length > 0 : props.filters[key] !== "")
      );
    };
    const setRowsPerPage = (numberOfRowsPerPage) => {
      rowsPerPage.value = numberOfRowsPerPage;
    };
    const setCurrentPage = (page) => {
      currentPage.value = page;
    };
    const sendFiltersTelemetry = (source) => {
      if (resettingFilters.value) {
        if (source !== "reset") {
          return;
        }
        setTimeout(() => resettingFilters.value = false, 1500);
      }
      const filters2 = filtersModel.value;
      const filtersSet = [];
      const filterValues = [];
      Object.keys(filters2).forEach((key) => {
        if (filters2[key]) {
          filtersSet.push(key);
          filterValues.push(key === "search" ? null : filters2[key]);
        }
      });
      telemetry.track(`User set filters in ${props.resourceKey} list`, {
        filters_set: filtersSet,
        filter_values: filterValues,
        [`${props.resourceKey}_total_in_view`]: props.resources.length,
        [`${props.resourceKey}_after_filtering`]: filteredAndSortedResources.value.length
      });
    };
    const onAddButtonClick = (e) => {
      emit("click:add", e);
    };
    const onUpdateFilters = (e) => {
      emit("update:filters", e);
    };
    const resetFilters = () => {
      Object.keys(filtersModel.value).forEach((key) => {
        filtersModel.value[key] = Array.isArray(filtersModel.value[key]) ? [] : "";
      });
      resettingFilters.value = true;
      sendFiltersTelemetry("reset");
      emit("update:filters", filtersModel.value);
    };
    const itemSize = () => {
      if ("itemSize" in props.typeProps) {
        return props.typeProps.itemSize;
      }
      return 0;
    };
    const getColumns = () => {
      if ("columns" in props.typeProps) {
        return props.typeProps.columns;
      }
      return {};
    };
    const sendSortingTelemetry = () => {
      telemetry.track(`User changed sorting in ${props.resourceKey} list`, {
        sorting: sortBy.value
      });
    };
    const onUpdateFiltersLength = (length) => {
      hasFilters.value = length > 0;
    };
    const onSearch = (s) => {
      filtersModel.value.search = s;
      emit("update:filters", filtersModel.value);
    };
    watch(
      () => props.filters,
      (value) => {
        filtersModel.value = value;
      }
    );
    watch(
      () => filtersModel.value.homeProject,
      () => {
        sendFiltersTelemetry("homeProject");
      }
    );
    watch(
      () => filtersModel.value.tags,
      () => {
        sendFiltersTelemetry("tags");
      }
    );
    watch(
      () => filtersModel.value.type,
      () => {
        sendFiltersTelemetry("type");
      }
    );
    watch(
      () => filtersModel.value.search,
      () => callDebounced(sendFiltersTelemetry, { debounceTime: 1e3, trailing: true }, "search")
    );
    watch(
      () => sortBy.value,
      (newValue) => {
        emit("sort", newValue);
        sendSortingTelemetry();
      }
    );
    watch(
      () => {
        var _a;
        return (_a = route == null ? void 0 : route.params) == null ? void 0 : _a.projectId;
      },
      () => {
        resetFilters();
      }
    );
    onMounted(async () => {
      await props.initialize();
      await nextTick();
      focusSearchInput();
      if (hasAppliedFilters()) {
        hasFilters.value = true;
      }
    });
    return {
      i18n,
      search: search2,
      usersStore,
      filterKeys,
      currentPage,
      rowsPerPage,
      filteredAndSortedResources,
      hasFilters,
      sortBy,
      resettingFilters,
      filtersModel,
      sendFiltersTelemetry,
      getColumns,
      itemSize,
      onAddButtonClick,
      onUpdateFiltersLength,
      onUpdateFilters,
      resetFilters,
      callDebounced,
      setCurrentPage,
      setRowsPerPage,
      onSearch
    };
  }
});
const filters = "_filters_eysnt_1";
const search = "_search_eysnt_16";
const listWrapper = "_listWrapper_eysnt_20";
const datatable = "_datatable_eysnt_30";
const style0 = {
  "filters-row": "_filters-row_eysnt_1",
  filters,
  search,
  listWrapper,
  "sort-and-filter": "_sort-and-filter_eysnt_26",
  datatable
};
const _hoisted_1 = {
  key: 0,
  class: "resource-list-loading"
};
const _hoisted_2 = { key: 0 };
const _hoisted_3 = {
  key: 0,
  class: "mt-xs"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_loading = resolveComponent("n8n-loading");
  const _component_n8n_action_box = resolveComponent("n8n-action-box");
  const _component_n8n_icon = resolveComponent("n8n-icon");
  const _component_n8n_input = resolveComponent("n8n-input");
  const _component_ResourceFiltersDropdown = __unplugin_components_0;
  const _component_n8n_option = resolveComponent("n8n-option");
  const _component_n8n_select = resolveComponent("n8n-select");
  const _component_n8n_button = resolveComponent("n8n-button");
  const _component_n8n_link = resolveComponent("n8n-link");
  const _component_n8n_info_tip = resolveComponent("n8n-info-tip");
  const _component_n8n_recycle_scroller = resolveComponent("n8n-recycle-scroller");
  const _component_n8n_datatable = resolveComponent("n8n-datatable");
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_PageViewLayoutList = __unplugin_components_1;
  const _component_PageViewLayout = __unplugin_components_2;
  return openBlock(), createBlock(_component_PageViewLayout, null, {
    header: withCtx(() => [
      renderSlot(_ctx.$slots, "header", {}, void 0, true)
    ]),
    default: withCtx(() => [
      _ctx.loading ? (openBlock(), createElementBlock("div", _hoisted_1, [
        createVNode(_component_n8n_loading, {
          rows: 25,
          "shrink-last": false
        })
      ])) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
        _ctx.resources.length === 0 ? (openBlock(), createElementBlock("div", _hoisted_2, [
          renderSlot(_ctx.$slots, "empty", {}, () => {
            var _a, _b;
            return [
              createVNode(_component_n8n_action_box, {
                "data-test-id": "empty-resources-list",
                emoji: "👋",
                heading: _ctx.i18n.baseText(
                  ((_a = _ctx.usersStore.currentUser) == null ? void 0 : _a.firstName) ? `${_ctx.resourceKey}.empty.heading` : `${_ctx.resourceKey}.empty.heading.userNotSetup`,
                  {
                    interpolate: { name: ((_b = _ctx.usersStore.currentUser) == null ? void 0 : _b.firstName) ?? "" }
                  }
                ),
                description: _ctx.i18n.baseText(`${_ctx.resourceKey}.empty.description`),
                "button-text": _ctx.i18n.baseText(`${_ctx.resourceKey}.empty.button`),
                "button-type": "secondary",
                "button-disabled": _ctx.disabled,
                "onClick:button": _ctx.onAddButtonClick
              }, {
                disabledButtonTooltip: withCtx(() => [
                  createTextVNode(toDisplayString(_ctx.i18n.baseText(`${_ctx.resourceKey}.empty.button.disabled.tooltip`)), 1)
                ]),
                _: 1
              }, 8, ["heading", "description", "button-text", "button-disabled", "onClick:button"])
            ];
          }, true)
        ])) : (openBlock(), createBlock(_component_PageViewLayoutList, {
          key: 1,
          overflow: _ctx.type !== "list"
        }, {
          header: withCtx(() => [
            createBaseVNode("div", {
              class: normalizeClass(_ctx.$style["filters-row"])
            }, [
              createBaseVNode("div", {
                class: normalizeClass(_ctx.$style.filters)
              }, [
                createVNode(_component_n8n_input, {
                  ref: "search",
                  "model-value": _ctx.filtersModel.search,
                  class: normalizeClass([_ctx.$style["search"], "mr-2xs"]),
                  placeholder: _ctx.i18n.baseText(`${_ctx.resourceKey}.search.placeholder`),
                  clearable: "",
                  "data-test-id": "resources-list-search",
                  "onUpdate:modelValue": _ctx.onSearch
                }, {
                  prefix: withCtx(() => [
                    createVNode(_component_n8n_icon, { icon: "search" })
                  ]),
                  _: 1
                }, 8, ["model-value", "class", "placeholder", "onUpdate:modelValue"]),
                _ctx.showFiltersDropdown ? (openBlock(), createBlock(_component_ResourceFiltersDropdown, {
                  key: 0,
                  keys: _ctx.filterKeys,
                  reset: _ctx.resetFilters,
                  "model-value": _ctx.filtersModel,
                  shareable: _ctx.shareable,
                  "onUpdate:modelValue": _ctx.onUpdateFilters,
                  "onUpdate:filtersLength": _ctx.onUpdateFiltersLength
                }, {
                  default: withCtx((resourceFiltersSlotProps) => [
                    renderSlot(_ctx.$slots, "filters", normalizeProps(guardReactiveProps(resourceFiltersSlotProps)), void 0, true)
                  ]),
                  _: 3
                }, 8, ["keys", "reset", "model-value", "shareable", "onUpdate:modelValue", "onUpdate:filtersLength"])) : createCommentVNode("", true),
                createBaseVNode("div", {
                  class: normalizeClass(_ctx.$style["sort-and-filter"])
                }, [
                  createVNode(_component_n8n_select, {
                    modelValue: _ctx.sortBy,
                    "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => _ctx.sortBy = $event),
                    "data-test-id": "resources-list-sort"
                  }, {
                    default: withCtx(() => [
                      (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.sortOptions, (sortOption) => {
                        return openBlock(), createBlock(_component_n8n_option, {
                          key: sortOption,
                          "data-test-id": "resources-list-sort-item",
                          value: sortOption,
                          label: _ctx.i18n.baseText(`${_ctx.resourceKey}.sort.${sortOption}`)
                        }, null, 8, ["value", "label"]);
                      }), 128))
                    ]),
                    _: 1
                  }, 8, ["modelValue"])
                ], 2)
              ], 2),
              renderSlot(_ctx.$slots, "add-button", { disabled: _ctx.disabled }, () => [
                createVNode(_component_n8n_button, {
                  size: "large",
                  disabled: _ctx.disabled,
                  "data-test-id": "resources-list-add",
                  onClick: _ctx.onAddButtonClick
                }, {
                  default: withCtx(() => [
                    createTextVNode(toDisplayString(_ctx.i18n.baseText(`${_ctx.resourceKey}.add`)), 1)
                  ]),
                  _: 1
                }, 8, ["disabled", "onClick"])
              ], true)
            ], 2),
            renderSlot(_ctx.$slots, "callout", {}, void 0, true),
            _ctx.showFiltersDropdown ? withDirectives((openBlock(), createElementBlock("div", _hoisted_3, [
              createVNode(_component_n8n_info_tip, { bold: false }, {
                default: withCtx(() => [
                  createTextVNode(toDisplayString(_ctx.i18n.baseText(`${_ctx.resourceKey}.filters.active`)) + " ", 1),
                  createVNode(_component_n8n_link, {
                    "data-test-id": "workflows-filter-reset",
                    size: "small",
                    onClick: _ctx.resetFilters
                  }, {
                    default: withCtx(() => [
                      createTextVNode(toDisplayString(_ctx.i18n.baseText(`${_ctx.resourceKey}.filters.active.reset`)), 1)
                    ]),
                    _: 1
                  }, 8, ["onClick"])
                ]),
                _: 1
              })
            ], 512)), [
              [vShow, _ctx.hasFilters]
            ]) : createCommentVNode("", true),
            _cache[1] || (_cache[1] = createBaseVNode("div", { class: "pb-xs" }, null, -1))
          ]),
          default: withCtx(() => [
            renderSlot(_ctx.$slots, "preamble", {}, void 0, true),
            _ctx.filteredAndSortedResources.length > 0 ? (openBlock(), createElementBlock("div", {
              key: 0,
              ref: "listWrapperRef",
              class: normalizeClass(_ctx.$style.listWrapper)
            }, [
              _ctx.type === "list" ? (openBlock(), createBlock(_component_n8n_recycle_scroller, {
                key: 0,
                "data-test-id": "resources-list",
                items: _ctx.filteredAndSortedResources,
                "item-size": _ctx.itemSize(),
                "item-key": "id"
              }, {
                default: withCtx(({ item, updateItemSize }) => [
                  renderSlot(_ctx.$slots, "default", {
                    data: item,
                    updateItemSize
                  }, void 0, true)
                ]),
                _: 3
              }, 8, ["items", "item-size"])) : createCommentVNode("", true),
              _ctx.type === "datatable" ? (openBlock(), createBlock(_component_n8n_datatable, {
                key: 1,
                "data-test-id": "resources-table",
                class: normalizeClass(_ctx.$style.datatable),
                columns: _ctx.getColumns(),
                rows: _ctx.filteredAndSortedResources,
                "current-page": _ctx.currentPage,
                "rows-per-page": _ctx.rowsPerPage,
                "onUpdate:currentPage": _ctx.setCurrentPage,
                "onUpdate:rowsPerPage": _ctx.setRowsPerPage
              }, {
                row: withCtx(({ columns, row }) => [
                  renderSlot(_ctx.$slots, "default", {
                    data: row,
                    columns
                  }, void 0, true)
                ]),
                _: 3
              }, 8, ["class", "columns", "rows", "current-page", "rows-per-page", "onUpdate:currentPage", "onUpdate:rowsPerPage"])) : createCommentVNode("", true)
            ], 2)) : (openBlock(), createBlock(_component_n8n_text, {
              key: 1,
              color: "text-base",
              size: "medium",
              "data-test-id": "resources-list-empty"
            }, {
              default: withCtx(() => [
                createTextVNode(toDisplayString(_ctx.i18n.baseText(`${_ctx.resourceKey}.noResults`)), 1)
              ]),
              _: 1
            })),
            renderSlot(_ctx.$slots, "postamble", {}, void 0, true)
          ]),
          _: 3
        }, 8, ["overflow"]))
      ], 64))
    ]),
    _: 3
  });
}
const cssModules = {
  "$style": style0
};
const ResourcesListLayout = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules], ["__scopeId", "data-v-b2fe67a5"]]);
export {
  ResourcesListLayout as R
};
