// n8n-editor-ui/dist/assets/SettingsLogStreamingView-D32eR0j1.js
import { d as defineComponent, al as useMessage, a8 as EnterpriseEditionFeature, dF as deepCopy, hH as defaultMessageEventBusDestinationOptions, e0 as mapStores, hI as useLogStreamingStore, ae as MODAL_CONFIRM, _ as _export_sfc, c as openBlock, e as createBlock, w as withCtx, j as createBaseVNode, i as createVNode, k as createTextVNode, t as toDisplayString, n as normalizeClass, l as resolveComponent, v as createEventBus, a3 as useDocumentTitle, dt as useCredentialsStore, K as useUIStore, T as useWorkflowsStore, m as useSettingsStore, an as hasPermission, eZ as v4, C as nextTick, hJ as LOG_STREAM_MODAL_KEY, h as createElementBlock, F as Fragment, f as createCommentVNode, as as withDirectives, z as renderList, c9 as resolveDirective } from "./index-TQ22MZub.js";
const DESTINATION_LIST_ITEM_ACTIONS = {
  OPEN: "open",
  DELETE: "delete"
};
const _sfc_main$1 = defineComponent({
  components: {},
  setup() {
    return {
      ...useMessage()
    };
  },
  data() {
    return {
      EnterpriseEditionFeature,
      nodeParameters: {}
    };
  },
  props: {
    eventBus: {
      type: Object
    },
    destination: {
      type: Object,
      required: true,
      default: deepCopy(defaultMessageEventBusDestinationOptions)
    },
    readonly: Boolean
  },
  mounted() {
    var _a;
    this.nodeParameters = Object.assign(
      deepCopy(defaultMessageEventBusDestinationOptions),
      this.destination
    );
    (_a = this.eventBus) == null ? void 0 : _a.on("destinationWasSaved", this.onDestinationWasSaved);
  },
  beforeUnmount() {
    var _a;
    (_a = this.eventBus) == null ? void 0 : _a.off("destinationWasSaved", this.onDestinationWasSaved);
  },
  computed: {
    ...mapStores(useLogStreamingStore),
    actions() {
      const actions = [
        {
          label: this.$locale.baseText("workflows.item.open"),
          value: DESTINATION_LIST_ITEM_ACTIONS.OPEN
        }
      ];
      if (!this.readonly) {
        actions.push({
          label: this.$locale.baseText("workflows.item.delete"),
          value: DESTINATION_LIST_ITEM_ACTIONS.DELETE
        });
      }
      return actions;
    },
    typeLabelName() {
      return `settings.log-streaming.${this.destination.__type}`;
    }
  },
  methods: {
    onDestinationWasSaved() {
      const updatedDestination = this.logStreamingStore.getDestination(this.destination.id);
      if (updatedDestination) {
        this.nodeParameters = Object.assign(
          deepCopy(defaultMessageEventBusDestinationOptions),
          this.destination
        );
      }
    },
    async onClick(event) {
      const cardActions2 = this.$refs.cardActions;
      const target = event.target;
      if (cardActions2 === target || (cardActions2 == null ? void 0 : cardActions2.contains(target)) || (target == null ? void 0 : target.contains(cardActions2))) {
        return;
      }
      this.$emit("edit", this.destination.id);
    },
    onEnabledSwitched(state) {
      this.nodeParameters.enabled = state;
      void this.saveDestination();
    },
    async saveDestination() {
      await this.logStreamingStore.saveDestination(this.nodeParameters);
    },
    async onAction(action) {
      if (action === DESTINATION_LIST_ITEM_ACTIONS.OPEN) {
        this.$emit("edit", this.destination.id);
      } else if (action === DESTINATION_LIST_ITEM_ACTIONS.DELETE) {
        const deleteConfirmed = await this.confirm(
          this.$locale.baseText("settings.log-streaming.destinationDelete.message", {
            interpolate: { destinationName: this.destination.label }
          }),
          this.$locale.baseText("settings.log-streaming.destinationDelete.headline"),
          {
            type: "warning",
            confirmButtonText: this.$locale.baseText(
              "settings.log-streaming.destinationDelete.confirmButtonText"
            ),
            cancelButtonText: this.$locale.baseText(
              "settings.log-streaming.destinationDelete.cancelButtonText"
            )
          }
        );
        if (deleteConfirmed !== MODAL_CONFIRM) {
          return;
        }
        this.$emit("remove", this.destination.id);
      }
    }
  }
});
const cardLink = "_cardLink_r6102_1";
const activeStatusText = "_activeStatusText_r6102_11";
const cardHeading = "_cardHeading_r6102_19";
const cardDescription = "_cardDescription_r6102_25";
const cardActions = "_cardActions_r6102_32";
const style0$1 = {
  cardLink,
  activeStatusText,
  cardHeading,
  cardDescription,
  cardActions
};
function _sfc_render$1(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_el_switch = resolveComponent("el-switch");
  const _component_n8n_action_toggle = resolveComponent("n8n-action-toggle");
  const _component_n8n_card = resolveComponent("n8n-card");
  return openBlock(), createBlock(_component_n8n_card, {
    class: normalizeClass(_ctx.$style.cardLink),
    "data-test-id": "destination-card",
    onClick: _ctx.onClick
  }, {
    header: withCtx(() => [
      createBaseVNode("div", null, [
        createVNode(_component_n8n_heading, {
          tag: "h2",
          bold: "",
          class: normalizeClass(_ctx.$style.cardHeading)
        }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(_ctx.destination.label), 1)
          ]),
          _: 1
        }, 8, ["class"]),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.cardDescription)
        }, [
          createVNode(_component_n8n_text, {
            color: "text-light",
            size: "small"
          }, {
            default: withCtx(() => [
              createBaseVNode("span", null, toDisplayString(_ctx.$locale.baseText(_ctx.typeLabelName)), 1)
            ]),
            _: 1
          })
        ], 2)
      ])
    ]),
    append: withCtx(() => [
      createBaseVNode("div", {
        ref: "cardActions",
        class: normalizeClass(_ctx.$style.cardActions)
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.activeStatusText),
          "data-test-id": "destination-activator-status"
        }, [
          _ctx.nodeParameters.enabled ? (openBlock(), createBlock(_component_n8n_text, {
            key: 0,
            color: "success",
            size: "small",
            bold: ""
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("workflowActivator.active")), 1)
            ]),
            _: 1
          })) : (openBlock(), createBlock(_component_n8n_text, {
            key: 1,
            color: "text-base",
            size: "small",
            bold: ""
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("workflowActivator.inactive")), 1)
            ]),
            _: 1
          }))
        ], 2),
        createVNode(_component_el_switch, {
          class: "mr-s",
          disabled: _ctx.readonly,
          "model-value": _ctx.nodeParameters.enabled,
          title: _ctx.nodeParameters.enabled ? _ctx.$locale.baseText("workflowActivator.deactivateWorkflow") : _ctx.$locale.baseText("workflowActivator.activateWorkflow"),
          "active-color": "#13ce66",
          "inactive-color": "#8899AA",
          "data-test-id": "workflow-activate-switch",
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => _ctx.onEnabledSwitched($event))
        }, null, 8, ["disabled", "model-value", "title"]),
        createVNode(_component_n8n_action_toggle, {
          actions: _ctx.actions,
          theme: "dark",
          onAction: _ctx.onAction
        }, null, 8, ["actions", "onAction"])
      ], 2)
    ]),
    _: 1
  }, 8, ["class", "onClick"]);
}
const cssModules$1 = {
  "$style": style0$1
};
const EventDestinationCard = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["render", _sfc_render$1], ["__cssModules", cssModules$1]]);
const _sfc_main = defineComponent({
  name: "SettingsLogStreamingView",
  components: {
    EventDestinationCard
  },
  props: {},
  data() {
    return {
      eventBus: createEventBus(),
      destinations: Array,
      disableLicense: false,
      allDestinations: [],
      documentTitle: useDocumentTitle()
    };
  },
  async mounted() {
    this.documentTitle.set(this.$locale.baseText("settings.log-streaming.heading"));
    if (!this.isLicensed) return;
    await this.credentialsStore.fetchCredentialTypes(false);
    await this.credentialsStore.fetchAllCredentials();
    this.uiStore.nodeViewInitialized = false;
    await this.getDestinationDataFromBackend();
    this.logStreamingStore.$onAction(({ name, after }) => {
      if (name === "removeDestination" || name === "updateDestination") {
        after(async () => {
          this.$forceUpdate();
        });
      }
    });
    this.eventBus.on("destinationWasSaved", this.onDestinationWasSaved);
    this.eventBus.on("remove", this.onRemove);
    this.eventBus.on("closing", this.onBusClosing);
  },
  beforeUnmount() {
    this.eventBus.off("destinationWasSaved", this.onDestinationWasSaved);
    this.eventBus.off("remove", this.onRemove);
    this.eventBus.off("closing", this.onBusClosing);
  },
  computed: {
    ...mapStores(
      useSettingsStore,
      useLogStreamingStore,
      useWorkflowsStore,
      useUIStore,
      useCredentialsStore
    ),
    sortedItemKeysByLabel() {
      var _a;
      const sortedKeys = [];
      for (const [key, value] of Object.entries(this.logStreamingStore.items)) {
        sortedKeys.push({ key, label: ((_a = value.destination) == null ? void 0 : _a.label) ?? "Destination" });
      }
      return sortedKeys.sort((a, b) => a.label.localeCompare(b.label));
    },
    environment() {
      return "production";
    },
    isLicensed() {
      if (this.disableLicense) return false;
      return this.settingsStore.isEnterpriseFeatureEnabled[EnterpriseEditionFeature.LogStreaming];
    },
    canManageLogStreaming() {
      return hasPermission(["rbac"], { rbac: { scope: "logStreaming:manage" } });
    }
  },
  methods: {
    onDestinationWasSaved() {
      this.$forceUpdate();
    },
    onBusClosing() {
      this.workflowsStore.removeAllNodes({ setStateDirty: false, removePinData: true });
      this.uiStore.stateIsDirty = false;
    },
    async getDestinationDataFromBackend() {
      this.logStreamingStore.clearEventNames();
      this.logStreamingStore.clearDestinations();
      this.allDestinations = [];
      const eventNamesData = await this.logStreamingStore.fetchEventNames();
      if (eventNamesData) {
        for (const eventName of eventNamesData) {
          this.logStreamingStore.addEventName(eventName);
        }
      }
      const destinationData = await this.logStreamingStore.fetchDestinations();
      if (destinationData) {
        for (const destination of destinationData) {
          this.logStreamingStore.addDestination(destination);
          this.allDestinations.push(destination);
        }
      }
      this.$forceUpdate();
    },
    goToUpgrade() {
      void this.uiStore.goToUpgrade("log-streaming", "upgrade-log-streaming");
    },
    storeHasItems() {
      return this.logStreamingStore.items && Object.keys(this.logStreamingStore.items).length > 0;
    },
    async addDestination() {
      const newDestination = deepCopy(defaultMessageEventBusDestinationOptions);
      newDestination.id = v4();
      this.logStreamingStore.addDestination(newDestination);
      await nextTick();
      this.uiStore.openModalWithData({
        name: LOG_STREAM_MODAL_KEY,
        data: {
          destination: newDestination,
          isNew: true,
          eventBus: this.eventBus
        }
      });
    },
    async onRemove(destinationId) {
      if (!destinationId) return;
      await this.logStreamingStore.deleteDestination(destinationId);
      const foundNode = this.workflowsStore.getNodeByName(destinationId);
      if (foundNode) {
        this.workflowsStore.removeNode(foundNode);
      }
    },
    async onEdit(destinationId) {
      if (!destinationId) return;
      const editDestination = this.logStreamingStore.getDestination(destinationId);
      if (editDestination) {
        this.uiStore.openModalWithData({
          name: LOG_STREAM_MODAL_KEY,
          data: {
            destination: editDestination,
            isNew: false,
            eventBus: this.eventBus
          }
        });
      }
    }
  }
});
const header = "_header_brzxt_1";
const destinationItem = "_destinationItem_brzxt_11";
const style0 = {
  header,
  destinationItem
};
const _hoisted_1 = { class: "mb-2xl" };
const _hoisted_2 = { class: "ml-m" };
const _hoisted_3 = { class: "mb-l" };
const _hoisted_4 = { class: "mt-m text-right" };
const _hoisted_5 = {
  key: 1,
  "data-test-id": "action-box-licensed"
};
const _hoisted_6 = {
  key: 0,
  class: "mb-l"
};
const _hoisted_7 = { "data-test-id": "action-box-unlicensed" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_el_switch = resolveComponent("el-switch");
  const _component_n8n_info_tip = resolveComponent("n8n-info-tip");
  const _component_EventDestinationCard = resolveComponent("EventDestinationCard");
  const _component_el_col = resolveComponent("el-col");
  const _component_el_row = resolveComponent("el-row");
  const _component_n8n_button = resolveComponent("n8n-button");
  const _component_n8n_action_box = resolveComponent("n8n-action-box");
  const _directive_n8n_html = resolveDirective("n8n-html");
  return openBlock(), createElementBlock("div", null, [
    createBaseVNode("div", {
      class: normalizeClass(_ctx.$style.header)
    }, [
      createBaseVNode("div", _hoisted_1, [
        createVNode(_component_n8n_heading, { size: "2xlarge" }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(_ctx.$locale.baseText(`settings.log-streaming.heading`)), 1)
          ]),
          _: 1
        }),
        _ctx.environment !== "production" ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
          createBaseVNode("strong", _hoisted_2, "Disable License (" + toDisplayString(_ctx.environment) + ") ", 1),
          createVNode(_component_el_switch, {
            modelValue: _ctx.disableLicense,
            "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => _ctx.disableLicense = $event),
            size: "large",
            "data-test-id": "disable-license-toggle"
          }, null, 8, ["modelValue"])
        ], 64)) : createCommentVNode("", true)
      ])
    ], 2),
    _ctx.isLicensed ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
      createBaseVNode("div", _hoisted_3, [
        createVNode(_component_n8n_info_tip, {
          theme: "info",
          type: "note"
        }, {
          default: withCtx(() => [
            withDirectives(createBaseVNode("span", null, null, 512), [
              [_directive_n8n_html, _ctx.$locale.baseText("settings.log-streaming.infoText")]
            ])
          ]),
          _: 1
        })
      ]),
      _ctx.storeHasItems() ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
        (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.sortedItemKeysByLabel, (item) => {
          return openBlock(), createBlock(_component_el_row, {
            key: item.key,
            gutter: 10,
            class: normalizeClass(_ctx.$style.destinationItem)
          }, {
            default: withCtx(() => {
              var _a;
              return [
                ((_a = _ctx.logStreamingStore.items[item.key]) == null ? void 0 : _a.destination) ? (openBlock(), createBlock(_component_el_col, { key: 0 }, {
                  default: withCtx(() => {
                    var _a2;
                    return [
                      createVNode(_component_EventDestinationCard, {
                        destination: (_a2 = _ctx.logStreamingStore.items[item.key]) == null ? void 0 : _a2.destination,
                        "event-bus": _ctx.eventBus,
                        readonly: !_ctx.canManageLogStreaming,
                        onRemove: ($event) => {
                          var _a3, _b;
                          return _ctx.onRemove((_b = (_a3 = _ctx.logStreamingStore.items[item.key]) == null ? void 0 : _a3.destination) == null ? void 0 : _b.id);
                        },
                        onEdit: ($event) => {
                          var _a3, _b;
                          return _ctx.onEdit((_b = (_a3 = _ctx.logStreamingStore.items[item.key]) == null ? void 0 : _a3.destination) == null ? void 0 : _b.id);
                        }
                      }, null, 8, ["destination", "event-bus", "readonly", "onRemove", "onEdit"])
                    ];
                  }),
                  _: 2
                }, 1024)) : createCommentVNode("", true)
              ];
            }),
            _: 2
          }, 1032, ["class"]);
        }), 128)),
        createBaseVNode("div", _hoisted_4, [
          _ctx.canManageLogStreaming ? (openBlock(), createBlock(_component_n8n_button, {
            key: 0,
            size: "large",
            onClick: _ctx.addDestination
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText(`settings.log-streaming.add`)), 1)
            ]),
            _: 1
          }, 8, ["onClick"])) : createCommentVNode("", true)
        ])
      ], 64)) : (openBlock(), createElementBlock("div", _hoisted_5, [
        createVNode(_component_n8n_action_box, {
          "button-text": _ctx.$locale.baseText(`settings.log-streaming.add`),
          "onClick:button": _ctx.addDestination
        }, {
          heading: withCtx(() => [
            withDirectives(createBaseVNode("span", null, null, 512), [
              [_directive_n8n_html, _ctx.$locale.baseText(`settings.log-streaming.addFirstTitle`)]
            ])
          ]),
          _: 1
        }, 8, ["button-text", "onClick:button"])
      ]))
    ], 64)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
      _ctx.$locale.baseText("settings.log-streaming.infoText") ? (openBlock(), createElementBlock("div", _hoisted_6, [
        createVNode(_component_n8n_info_tip, {
          theme: "info",
          type: "note"
        }, {
          default: withCtx(() => [
            withDirectives(createBaseVNode("span", null, null, 512), [
              [_directive_n8n_html, _ctx.$locale.baseText("settings.log-streaming.infoText")]
            ])
          ]),
          _: 1
        })
      ])) : createCommentVNode("", true),
      createBaseVNode("div", _hoisted_7, [
        createVNode(_component_n8n_action_box, {
          description: _ctx.$locale.baseText("settings.log-streaming.actionBox.description"),
          "button-text": _ctx.$locale.baseText("settings.log-streaming.actionBox.button"),
          "onClick:button": _ctx.goToUpgrade
        }, {
          heading: withCtx(() => [
            withDirectives(createBaseVNode("span", null, null, 512), [
              [_directive_n8n_html, _ctx.$locale.baseText("settings.log-streaming.actionBox.title")]
            ])
          ]),
          _: 1
        }, 8, ["description", "button-text", "onClick:button"])
      ])
    ], 64))
  ]);
}
const cssModules = {
  "$style": style0
};
const SettingsLogStreamingView = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules]]);
export {
  SettingsLogStreamingView as default
};
