// n8n-editor-ui/dist/assets/SettingsApiView-DnaOHGOU.js
import { d as defineComponent, ep as CopyInput, a as useToast, al as useMessage, K as useUIStore, a3 as useDocumentTitle, hG as DOCS_DOMAIN, e0 as mapStores, az as useCloudPlanStore, u as useUsersStore, m as useSettingsStore, Z as useRootStore, ae as MODAL_CONFIRM, _ as _export_sfc, c as openBlock, h as createElementBlock, j as createBaseVNode, i as createVNode, w as withCtx, k as createTextVNode, t as toDisplayString, n as normalizeClass, e as createBlock, f as createCommentVNode, l as resolveComponent } from "./index-TQ22MZub.js";
const _sfc_main = defineComponent({
  name: "SettingsApiView",
  components: {
    CopyInput
  },
  setup() {
    return {
      ...useToast(),
      ...useMessage(),
      ...useUIStore(),
      documentTitle: useDocumentTitle()
    };
  },
  data() {
    return {
      loading: false,
      mounted: false,
      apiKeys: [],
      swaggerUIEnabled: false,
      apiDocsURL: ""
    };
  },
  mounted() {
    this.documentTitle.set(this.$locale.baseText("settings.api"));
    if (!this.isPublicApiEnabled) return;
    void this.getApiKeys();
    const baseUrl = this.rootStore.baseUrl;
    const apiPath = this.settingsStore.publicApiPath;
    const latestVersion = this.settingsStore.publicApiLatestVersion;
    this.swaggerUIEnabled = this.settingsStore.isSwaggerUIEnabled;
    this.apiDocsURL = this.swaggerUIEnabled ? `${baseUrl}${apiPath}/v${latestVersion}/docs` : `https://${DOCS_DOMAIN}/api/api-reference/`;
  },
  computed: {
    ...mapStores(useRootStore, useSettingsStore, useUsersStore, useCloudPlanStore, useUIStore),
    currentUser() {
      return this.usersStore.currentUser;
    },
    isTrialing() {
      return this.cloudPlanStore.userIsTrialing;
    },
    isLoadingCloudPlans() {
      return this.cloudPlanStore.state.loadingPlan;
    },
    isPublicApiEnabled() {
      return this.settingsStore.isPublicApiEnabled;
    },
    isRedactedApiKey() {
      if (!this.apiKeys) return false;
      return this.apiKeys[0].apiKey.includes("*");
    }
  },
  methods: {
    onUpgrade() {
      void this.uiStore.goToUpgrade("settings-n8n-api", "upgrade-api", "redirect");
    },
    async showDeleteModal() {
      const confirmed = await this.confirm(
        this.$locale.baseText("settings.api.delete.description"),
        this.$locale.baseText("settings.api.delete.title"),
        {
          confirmButtonText: this.$locale.baseText("settings.api.delete.button"),
          cancelButtonText: this.$locale.baseText("generic.cancel")
        }
      );
      if (confirmed === MODAL_CONFIRM) {
        await this.deleteApiKey();
      }
    },
    async getApiKeys() {
      try {
        this.apiKeys = await this.settingsStore.getApiKeys();
      } catch (error) {
        this.showError(error, this.$locale.baseText("settings.api.view.error"));
      } finally {
        this.mounted = true;
      }
    },
    async createApiKey() {
      this.loading = true;
      try {
        const newApiKey = await this.settingsStore.createApiKey();
        this.apiKeys.push(newApiKey);
      } catch (error) {
        this.showError(error, this.$locale.baseText("settings.api.create.error"));
      } finally {
        this.loading = false;
        this.$telemetry.track("User clicked create API key button");
      }
    },
    async deleteApiKey() {
      try {
        await this.settingsStore.deleteApiKey(this.apiKeys[0].id);
        this.showMessage({
          title: this.$locale.baseText("settings.api.delete.toast"),
          type: "success"
        });
        this.apiKeys = [];
      } catch (error) {
        this.showError(error, this.$locale.baseText("settings.api.delete.error"));
      } finally {
        this.$telemetry.track("User clicked delete API key button");
      }
    },
    onCopy() {
      this.$telemetry.track("User clicked copy API key button");
    }
  }
});
const container = "_container_1xc1f_1";
const header = "_header_1xc1f_5";
const card = "_card_1xc1f_14";
const hint = "_hint_1xc1f_25";
const style0 = {
  container,
  header,
  card,
  "delete": "_delete_1xc1f_18",
  hint
};
const _hoisted_1 = { style: { fontSize: "var(--font-size-s)", color: "var(--color-text-light)" } };
const _hoisted_2 = { key: 0 };
const _hoisted_3 = { class: "mb-s" };
const _hoisted_4 = ["textContent"];
const _hoisted_5 = ["textContent"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_i18n_t = resolveComponent("i18n-t");
  const _component_n8n_info_tip = resolveComponent("n8n-info-tip");
  const _component_n8n_link = resolveComponent("n8n-link");
  const _component_CopyInput = CopyInput;
  const _component_n8n_card = resolveComponent("n8n-card");
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_n8n_action_box = resolveComponent("n8n-action-box");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.$style.container)
  }, [
    createBaseVNode("div", {
      class: normalizeClass(_ctx.$style.header)
    }, [
      createVNode(_component_n8n_heading, { size: "2xlarge" }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("settings.api")) + " ", 1),
          createBaseVNode("span", _hoisted_1, " (" + toDisplayString(_ctx.$locale.baseText("generic.beta")) + ") ", 1)
        ]),
        _: 1
      })
    ], 2),
    _ctx.apiKeys.length ? (openBlock(), createElementBlock("div", _hoisted_2, [
      createBaseVNode("p", _hoisted_3, [
        createVNode(_component_n8n_info_tip, { bold: false }, {
          default: withCtx(() => [
            createVNode(_component_i18n_t, {
              keypath: "settings.api.view.info",
              tag: "span"
            }, {
              apiAction: withCtx(() => [
                createBaseVNode("a", {
                  href: "https://docs.n8n.io/api",
                  target: "_blank",
                  textContent: toDisplayString(_ctx.$locale.baseText("settings.api.view.info.api"))
                }, null, 8, _hoisted_4)
              ]),
              webhookAction: withCtx(() => [
                createBaseVNode("a", {
                  href: "https://docs.n8n.io/integrations/core-nodes/n8n-nodes-base.webhook/",
                  target: "_blank",
                  textContent: toDisplayString(_ctx.$locale.baseText("settings.api.view.info.webhook"))
                }, null, 8, _hoisted_5)
              ]),
              _: 1
            })
          ]),
          _: 1
        })
      ]),
      createVNode(_component_n8n_card, {
        class: normalizeClass(["mb-4xs", _ctx.$style.card])
      }, {
        default: withCtx(() => [
          createBaseVNode("span", {
            class: normalizeClass(_ctx.$style.delete)
          }, [
            createVNode(_component_n8n_link, {
              bold: true,
              onClick: _ctx.showDeleteModal
            }, {
              default: withCtx(() => [
                createTextVNode(toDisplayString(_ctx.$locale.baseText("generic.delete")), 1)
              ]),
              _: 1
            }, 8, ["onClick"])
          ], 2),
          createBaseVNode("div", null, [
            createVNode(_component_CopyInput, {
              label: _ctx.apiKeys[0].label,
              value: _ctx.apiKeys[0].apiKey,
              "copy-button-text": _ctx.$locale.baseText("generic.clickToCopy"),
              "toast-title": _ctx.$locale.baseText("settings.api.view.copy.toast"),
              "redact-value": true,
              "disable-copy": _ctx.isRedactedApiKey,
              hint: !_ctx.isRedactedApiKey ? _ctx.$locale.baseText("settings.api.view.copy") : "",
              onCopy: _ctx.onCopy
            }, null, 8, ["label", "value", "copy-button-text", "toast-title", "disable-copy", "hint", "onCopy"])
          ])
        ]),
        _: 1
      }, 8, ["class"]),
      createBaseVNode("div", {
        class: normalizeClass(_ctx.$style.hint)
      }, [
        createVNode(_component_n8n_text, { size: "small" }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(_ctx.$locale.baseText(`settings.api.view.${_ctx.swaggerUIEnabled ? "tryapi" : "more-details"}`)), 1)
          ]),
          _: 1
        }),
        _cache[0] || (_cache[0] = createTextVNode(" " + toDisplayString(" ") + " ")),
        createVNode(_component_n8n_link, {
          to: _ctx.apiDocsURL,
          "new-window": true,
          size: "small"
        }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(_ctx.$locale.baseText(
              `settings.api.view.${_ctx.swaggerUIEnabled ? "apiPlayground" : "external-docs"}`
            )), 1)
          ]),
          _: 1
        }, 8, ["to"])
      ], 2)
    ])) : !_ctx.isPublicApiEnabled && _ctx.isTrialing ? (openBlock(), createBlock(_component_n8n_action_box, {
      key: 1,
      "data-test-id": "public-api-upgrade-cta",
      heading: _ctx.$locale.baseText("settings.api.trial.upgradePlan.title"),
      description: _ctx.$locale.baseText("settings.api.trial.upgradePlan.description"),
      "button-text": _ctx.$locale.baseText("settings.api.trial.upgradePlan.cta"),
      "onClick:button": _ctx.onUpgrade
    }, null, 8, ["heading", "description", "button-text", "onClick:button"])) : _ctx.mounted && !_ctx.isLoadingCloudPlans ? (openBlock(), createBlock(_component_n8n_action_box, {
      key: 2,
      "button-text": _ctx.$locale.baseText(
        _ctx.loading ? "settings.api.create.button.loading" : "settings.api.create.button"
      ),
      description: _ctx.$locale.baseText("settings.api.create.description"),
      "onClick:button": _ctx.createApiKey
    }, null, 8, ["button-text", "description", "onClick:button"])) : createCommentVNode("", true)
  ], 2);
}
const cssModules = {
  "$style": style0
};
const SettingsApiView = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules]]);
export {
  SettingsApiView as default
};
