// n8n-editor-ui/dist/assets/SettingsFakeDoorView-DJJ9xpvu.js
import { d as defineComponent, hK as __unplugin_components_0, e0 as mapStores, K as useUIStore, _ as _export_sfc, c as openBlock, e as createBlock } from "./index-TQ22MZub.js";
const _sfc_main = defineComponent({
  name: "SettingsFakeDoorView",
  components: {
    FeatureComingSoon: __unplugin_components_0
  },
  props: {
    featureId: {
      type: String,
      required: true
    }
  },
  computed: {
    ...mapStores(useUIStore),
    featureInfo() {
      return this.uiStore.fakeDoorsById[this.featureId];
    }
  },
  methods: {
    openLinkPage() {
      if (this.featureInfo) {
        window.open(this.featureInfo.linkURL, "_blank");
      }
    }
  }
});
const header = "_header_fa9t3_1";
const style0 = {
  header
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_FeatureComingSoon = __unplugin_components_0;
  return openBlock(), createBlock(_component_FeatureComingSoon, {
    "feature-id": _ctx.featureId,
    "show-title": ""
  }, null, 8, ["feature-id"]);
}
const cssModules = {
  "$style": style0
};
const SettingsFakeDoorView = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules]]);
export {
  SettingsFakeDoorView as default
};
