// n8n-editor-ui/dist/assets/SettingsCommunityNodesView-DG_uJUfX.js
import { d as defineComponent, hD as COMMUNITY_PACKAGE_MANAGE_ACTIONS, e0 as mapStores, K as useUIStore, d5 as NPM_PACKAGE_DOCS_BASE_URL, _ as _export_sfc, c as openBlock, h as createElementBlock, n as normalizeClass, i as createVNode, j as createBaseVNode, w as withCtx, k as createTextVNode, t as toDisplayString, F as Fragment, z as renderList, f as createCommentVNode, e as createBlock, l as resolveComponent, r as ref, b as useRouter, a as useToast, a3 as useDocumentTitle, hE as useCommunityNodesStore, p as computed, ar as onBeforeMount, o as onMounted, y as onBeforeUnmount, A as unref, g as useI18n, ak as useTelemetry, hF as COMMUNITY_PACKAGE_INSTALL_MODAL_KEY, aF as useExternalHooks, d7 as COMMUNITY_NODES_INSTALLATION_DOCS_URL } from "./index-TQ22MZub.js";
import { u as usePushConnection } from "./usePushConnection-Bg_ZQZs0.js";
import { u as usePushConnectionStore } from "./pushConnection.store-DBI9rig9.js";
const _sfc_main$1 = defineComponent({
  name: "CommunityPackageCard",
  props: {
    communityPackage: {
      type: Object,
      required: false,
      default: null
    },
    loading: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      packageActions: [
        {
          label: this.$locale.baseText("settings.communityNodes.viewDocsAction.label"),
          value: COMMUNITY_PACKAGE_MANAGE_ACTIONS.VIEW_DOCS,
          type: "external-link"
        },
        {
          label: this.$locale.baseText("settings.communityNodes.uninstallAction.label"),
          value: COMMUNITY_PACKAGE_MANAGE_ACTIONS.UNINSTALL
        }
      ]
    };
  },
  computed: {
    ...mapStores(useUIStore)
  },
  methods: {
    async onAction(value) {
      if (!this.communityPackage) return;
      switch (value) {
        case COMMUNITY_PACKAGE_MANAGE_ACTIONS.VIEW_DOCS:
          this.$telemetry.track("user clicked to browse the cnr package documentation", {
            package_name: this.communityPackage.packageName,
            package_version: this.communityPackage.installedVersion
          });
          window.open(`${NPM_PACKAGE_DOCS_BASE_URL}${this.communityPackage.packageName}`, "_blank");
          break;
        case COMMUNITY_PACKAGE_MANAGE_ACTIONS.UNINSTALL:
          this.uiStore.openCommunityPackageUninstallConfirmModal(this.communityPackage.packageName);
          break;
      }
    },
    onUpdateClick() {
      if (!this.communityPackage) return;
      this.uiStore.openCommunityPackageUpdateConfirmModal(this.communityPackage.packageName);
    }
  }
});
const cardContainer = "_cardContainer_1cg0i_1";
const packageCard = "_packageCard_1cg0i_9";
const cardSkeleton = "_cardSkeleton_1cg0i_10";
const loader = "_loader_1cg0i_24";
const cardInfoContainer = "_cardInfoContainer_1cg0i_35";
const cardTitle = "_cardTitle_1cg0i_40";
const cardSubtitle = "_cardSubtitle_1cg0i_47";
const cardControlsContainer = "_cardControlsContainer_1cg0i_52";
const cardActions = "_cardActions_1cg0i_58";
const style0$1 = {
  cardContainer,
  packageCard,
  cardSkeleton,
  loader,
  cardInfoContainer,
  cardTitle,
  cardSubtitle,
  cardControlsContainer,
  cardActions
};
const _hoisted_1 = { key: 0 };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_loading = resolveComponent("n8n-loading");
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_n8n_icon = resolveComponent("n8n-icon");
  const _component_n8n_tooltip = resolveComponent("n8n-tooltip");
  const _component_n8n_button = resolveComponent("n8n-button");
  const _component_n8n_action_toggle = resolveComponent("n8n-action-toggle");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.$style.cardContainer),
    "data-test-id": "community-package-card"
  }, [
    _ctx.loading ? (openBlock(), createElementBlock("div", {
      key: 0,
      class: normalizeClass(_ctx.$style.cardSkeleton)
    }, [
      createVNode(_component_n8n_loading, {
        class: normalizeClass(_ctx.$style.loader),
        variant: "p",
        rows: 1
      }, null, 8, ["class"]),
      createVNode(_component_n8n_loading, {
        class: normalizeClass(_ctx.$style.loader),
        variant: "p",
        rows: 1
      }, null, 8, ["class"])
    ], 2)) : _ctx.communityPackage ? (openBlock(), createElementBlock("div", {
      key: 1,
      class: normalizeClass(_ctx.$style.packageCard)
    }, [
      createBaseVNode("div", {
        class: normalizeClass(_ctx.$style.cardInfoContainer)
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.cardTitle)
        }, [
          createVNode(_component_n8n_text, {
            bold: true,
            size: "large"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.communityPackage.packageName), 1)
            ]),
            _: 1
          })
        ], 2),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.cardSubtitle)
        }, [
          createVNode(_component_n8n_text, {
            bold: true,
            size: "small",
            color: "text-light"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("settings.communityNodes.packageNodes.label", {
                adjustToNumber: _ctx.communityPackage.installedNodes.length
              })) + ":  ", 1)
            ]),
            _: 1
          }),
          createVNode(_component_n8n_text, {
            size: "small",
            color: "text-light"
          }, {
            default: withCtx(() => [
              (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.communityPackage.installedNodes, (node, index) => {
                return openBlock(), createElementBlock("span", {
                  key: node.name
                }, [
                  createTextVNode(toDisplayString(node.name), 1),
                  index != _ctx.communityPackage.installedNodes.length - 1 ? (openBlock(), createElementBlock("span", _hoisted_1, ",")) : createCommentVNode("", true)
                ]);
              }), 128))
            ]),
            _: 1
          })
        ], 2)
      ], 2),
      createBaseVNode("div", {
        class: normalizeClass(_ctx.$style.cardControlsContainer)
      }, [
        createVNode(_component_n8n_text, {
          bold: true,
          size: "large",
          color: "text-light"
        }, {
          default: withCtx(() => [
            createTextVNode(" v" + toDisplayString(_ctx.communityPackage.installedVersion), 1)
          ]),
          _: 1
        }),
        _ctx.communityPackage.failedLoading === true ? (openBlock(), createBlock(_component_n8n_tooltip, {
          key: 0,
          placement: "top"
        }, {
          content: withCtx(() => [
            createBaseVNode("div", null, toDisplayString(_ctx.$locale.baseText("settings.communityNodes.failedToLoad.tooltip")), 1)
          ]),
          default: withCtx(() => [
            createVNode(_component_n8n_icon, {
              icon: "exclamation-triangle",
              color: "danger",
              size: "large"
            })
          ]),
          _: 1
        })) : _ctx.communityPackage.updateAvailable ? (openBlock(), createBlock(_component_n8n_tooltip, {
          key: 1,
          placement: "top"
        }, {
          content: withCtx(() => [
            createBaseVNode("div", null, toDisplayString(_ctx.$locale.baseText("settings.communityNodes.updateAvailable.tooltip")), 1)
          ]),
          default: withCtx(() => [
            createVNode(_component_n8n_button, {
              outline: "",
              label: "Update",
              onClick: _ctx.onUpdateClick
            }, null, 8, ["onClick"])
          ]),
          _: 1
        })) : (openBlock(), createBlock(_component_n8n_tooltip, {
          key: 2,
          placement: "top"
        }, {
          content: withCtx(() => [
            createBaseVNode("div", null, toDisplayString(_ctx.$locale.baseText("settings.communityNodes.upToDate.tooltip")), 1)
          ]),
          default: withCtx(() => [
            createVNode(_component_n8n_icon, {
              icon: "check-circle",
              color: "text-light",
              size: "large"
            })
          ]),
          _: 1
        })),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.cardActions)
        }, [
          createVNode(_component_n8n_action_toggle, {
            actions: _ctx.packageActions,
            onAction: _ctx.onAction
          }, null, 8, ["actions", "onAction"])
        ], 2)
      ], 2)
    ], 2)) : createCommentVNode("", true)
  ], 2);
}
const cssModules$1 = {
  "$style": style0$1
};
const CommunityPackageCard = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["render", _sfc_render], ["__cssModules", cssModules$1]]);
const PACKAGE_COUNT_THRESHOLD = 31;
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "SettingsCommunityNodesView",
  setup(__props) {
    const loading = ref(false);
    const router = useRouter();
    const pushConnection = usePushConnection({ router });
    const pushStore = usePushConnectionStore();
    const externalHooks = useExternalHooks();
    const i18n = useI18n();
    const telemetry = useTelemetry();
    const toast = useToast();
    const documentTitle = useDocumentTitle();
    const communityNodesStore = useCommunityNodesStore();
    const uiStore = useUIStore();
    const getEmptyStateDescription = computed(() => {
      const packageCount = communityNodesStore.availablePackageCount;
      return packageCount < PACKAGE_COUNT_THRESHOLD ? i18n.baseText("settings.communityNodes.empty.description.no-packages", {
        interpolate: {
          docURL: COMMUNITY_NODES_INSTALLATION_DOCS_URL
        }
      }) : i18n.baseText("settings.communityNodes.empty.description", {
        interpolate: {
          docURL: COMMUNITY_NODES_INSTALLATION_DOCS_URL,
          count: (Math.floor(packageCount / 10) * 10).toString()
        }
      });
    });
    const getEmptyStateButtonText = computed(
      () => i18n.baseText("settings.communityNodes.empty.installPackageLabel")
    );
    const actionBoxConfig = computed(() => {
      return {
        calloutText: "",
        calloutTheme: "",
        hideButton: false
      };
    });
    const onClickEmptyStateButton = () => {
      openInstallModal();
    };
    const openInstallModal = () => {
      const telemetryPayload = {
        is_empty_state: communityNodesStore.getInstalledPackages.length === 0
      };
      telemetry.track("user clicked cnr install button", telemetryPayload);
      void externalHooks.run("settingsCommunityNodesView.openInstallModal", telemetryPayload);
      uiStore.openModal(COMMUNITY_PACKAGE_INSTALL_MODAL_KEY);
    };
    onBeforeMount(() => {
      pushConnection.initialize();
      pushStore.pushConnect();
    });
    onMounted(async () => {
      documentTitle.set(i18n.baseText("settings.communityNodes"));
      try {
        loading.value = true;
        await communityNodesStore.fetchInstalledPackages();
        const installedPackages = communityNodesStore.getInstalledPackages;
        const packagesToUpdate = installedPackages.filter(
          (p) => p.updateAvailable
        );
        telemetry.track("user viewed cnr settings page", {
          num_of_packages_installed: installedPackages.length,
          installed_packages: installedPackages.map((p) => {
            return {
              package_name: p.packageName,
              package_version: p.installedVersion,
              package_nodes: p.installedNodes.map((node) => `${node.name}-v${node.latestVersion}`),
              is_update_available: p.updateAvailable !== void 0
            };
          }),
          packages_to_update: packagesToUpdate.map((p) => {
            return {
              package_name: p.packageName,
              package_version_current: p.installedVersion,
              package_version_available: p.updateAvailable
            };
          }),
          number_of_updates_available: packagesToUpdate.length
        });
      } catch (error) {
        toast.showError(
          error,
          i18n.baseText("settings.communityNodes.fetchError.title"),
          i18n.baseText("settings.communityNodes.fetchError.message")
        );
      } finally {
        loading.value = false;
      }
      try {
        await communityNodesStore.fetchAvailableCommunityPackageCount();
      } finally {
        loading.value = false;
      }
    });
    onBeforeUnmount(() => {
      pushStore.pushDisconnect();
      pushConnection.terminate();
    });
    return (_ctx, _cache) => {
      const _component_n8n_heading = resolveComponent("n8n-heading");
      const _component_n8n_button = resolveComponent("n8n-button");
      const _component_n8n_action_box = resolveComponent("n8n-action-box");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.container)
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.headingContainer)
        }, [
          createVNode(_component_n8n_heading, { size: "2xlarge" }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(unref(i18n).baseText("settings.communityNodes")), 1)
            ]),
            _: 1
          }),
          unref(communityNodesStore).getInstalledPackages.length > 0 && !loading.value ? (openBlock(), createBlock(_component_n8n_button, {
            key: 0,
            label: unref(i18n).baseText("settings.communityNodes.installModal.installButton.label"),
            size: "large",
            onClick: openInstallModal
          }, null, 8, ["label"])) : createCommentVNode("", true)
        ], 2),
        loading.value ? (openBlock(), createElementBlock("div", {
          key: 0,
          class: normalizeClass(_ctx.$style.cardsContainer)
        }, [
          (openBlock(), createElementBlock(Fragment, null, renderList(2, (n) => {
            return createVNode(CommunityPackageCard, {
              key: "index-" + n,
              loading: true
            });
          }), 64))
        ], 2)) : unref(communityNodesStore).getInstalledPackages.length === 0 ? (openBlock(), createElementBlock("div", {
          key: 1,
          class: normalizeClass(_ctx.$style.actionBoxContainer)
        }, [
          createVNode(_component_n8n_action_box, {
            heading: unref(i18n).baseText("settings.communityNodes.empty.title"),
            description: getEmptyStateDescription.value,
            "button-text": getEmptyStateButtonText.value,
            "callout-text": actionBoxConfig.value.calloutText,
            "callout-theme": actionBoxConfig.value.calloutTheme,
            "onClick:button": onClickEmptyStateButton
          }, null, 8, ["heading", "description", "button-text", "callout-text", "callout-theme"])
        ], 2)) : (openBlock(), createElementBlock("div", {
          key: 2,
          class: normalizeClass(_ctx.$style.cardsContainer)
        }, [
          (openBlock(true), createElementBlock(Fragment, null, renderList(unref(communityNodesStore).getInstalledPackages, (communityPackage) => {
            return openBlock(), createBlock(CommunityPackageCard, {
              key: communityPackage.packageName,
              "community-package": communityPackage
            }, null, 8, ["community-package"]);
          }), 128))
        ], 2))
      ], 2);
    };
  }
});
const container = "_container_18kse_1";
const headingContainer = "_headingContainer_18kse_9";
const loadingContainer = "_loadingContainer_18kse_14";
const actionBoxContainer = "_actionBoxContainer_18kse_19";
const cardsContainer = "_cardsContainer_18kse_23";
const style0 = {
  container,
  headingContainer,
  loadingContainer,
  actionBoxContainer,
  cardsContainer
};
const cssModules = {
  "$style": style0
};
const SettingsCommunityNodesView = /* @__PURE__ */ _export_sfc(_sfc_main, [["__cssModules", cssModules]]);
export {
  SettingsCommunityNodesView as default
};
