// n8n-editor-ui/dist/assets/TemplateList-DqeiM-sc.js
import { d as defineComponent, d3 as _sfc_main$3, hT as filterTemplateNodes, _ as _export_sfc, c as openBlock, h as createElementBlock, F as Fragment, z as renderList, n as normalizeClass, i as createVNode, t as toDisplayString, f as createCommentVNode, hR as _sfc_main$4, hS as abbreviateNumber, w as withCtx, k as createTextVNode, e as createBlock, I as withModifiers, l as resolveComponent, j as createBaseVNode } from "./index-TQ22MZub.js";
const _sfc_main$2 = defineComponent({
  name: "NodeList",
  components: {
    NodeIcon: _sfc_main$3
  },
  props: {
    nodes: {
      type: Array
    },
    limit: {
      type: Number,
      default: 4
    },
    size: {
      type: String,
      default: "sm"
    }
  },
  computed: {
    filteredCoreNodes() {
      return filterTemplateNodes(this.nodes);
    },
    hiddenNodes() {
      return this.filteredCoreNodes.length - this.countNodesToBeSliced(this.filteredCoreNodes);
    },
    slicedNodes() {
      return this.filteredCoreNodes.slice(0, this.countNodesToBeSliced(this.filteredCoreNodes));
    }
  },
  methods: {
    countNodesToBeSliced(nodes2) {
      if (nodes2.length > this.limit) {
        return this.limit - 1;
      } else {
        return this.limit;
      }
    }
  }
});
const list = "_list_10uzg_1";
const container = "_container_10uzg_9";
const sm = "_sm_10uzg_14";
const md = "_md_10uzg_18";
const button$1 = "_button_10uzg_22";
const buttonSm = "_buttonSm_10uzg_36";
const buttonMd = "_buttonMd_10uzg_43";
const style0$2 = {
  list,
  container,
  sm,
  md,
  button: button$1,
  buttonSm,
  buttonMd
};
function _sfc_render$2(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_NodeIcon = _sfc_main$3;
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.$style.list)
  }, [
    (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.slicedNodes, (node) => {
      return openBlock(), createElementBlock("div", {
        key: node.name,
        class: normalizeClass([_ctx.$style.container, _ctx.$style[_ctx.size]])
      }, [
        createVNode(_component_NodeIcon, {
          "node-type": node,
          size: _ctx.size === "md" ? 24 : 18,
          "show-tooltip": true
        }, null, 8, ["node-type", "size"])
      ], 2);
    }), 128)),
    _ctx.filteredCoreNodes.length > _ctx.limit + 1 ? (openBlock(), createElementBlock("div", {
      key: 0,
      class: normalizeClass([_ctx.$style.button, _ctx.size === "md" ? _ctx.$style.buttonMd : _ctx.$style.buttonSm])
    }, " +" + toDisplayString(_ctx.hiddenNodes), 3)) : createCommentVNode("", true)
  ], 2);
}
const cssModules$2 = {
  "$style": style0$2
};
const __unplugin_components_0$1 = /* @__PURE__ */ _export_sfc(_sfc_main$2, [["render", _sfc_render$2], ["__cssModules", cssModules$2]]);
const _sfc_main$1 = defineComponent({
  name: "TemplateCard",
  components: {
    TimeAgo: _sfc_main$4,
    NodeList: __unplugin_components_0$1
  },
  props: {
    workflow: {
      type: Object
    },
    lastItem: {
      type: Boolean,
      default: false
    },
    firstItem: {
      type: Boolean,
      default: false
    },
    useWorkflowButton: {
      type: Boolean
    },
    loading: {
      type: Boolean
    },
    simpleView: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      nodesToBeShown: 5
    };
  },
  methods: {
    filterTemplateNodes,
    abbreviateNumber,
    countNodesToBeSliced(nodes2) {
      if (nodes2.length > this.nodesToBeShown) {
        return this.nodesToBeShown - 1;
      } else {
        return this.nodesToBeShown;
      }
    },
    onUseWorkflowClick(e) {
      this.$emit("useWorkflow", e);
    },
    onCardClick(e) {
      this.$emit("click", e);
    }
  }
});
const nodes$1 = "_nodes_psqnp_1";
const icon = "_icon_psqnp_8";
const card = "_card_psqnp_12";
const hideOnHover = "_hideOnHover_psqnp_24";
const buttonContainer = "_buttonContainer_psqnp_27";
const loaded = "_loaded_psqnp_38";
const first = "_first_psqnp_42";
const last = "_last_psqnp_48";
const content = "_content_psqnp_53";
const line = "_line_psqnp_58";
const loading = "_loading_psqnp_64";
const nodesContainer = "_nodesContainer_psqnp_69";
const style0$1 = {
  nodes: nodes$1,
  icon,
  card,
  hideOnHover,
  buttonContainer,
  loaded,
  first,
  last,
  content,
  line,
  loading,
  nodesContainer
};
const _hoisted_1$1 = { key: 1 };
const _hoisted_2$1 = { key: 0 };
function _sfc_render$1(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_loading = resolveComponent("n8n-loading");
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_font_awesome_icon = resolveComponent("font-awesome-icon");
  const _component_n8n_text = resolveComponent("n8n-text");
  const _component_TimeAgo = _sfc_main$4;
  const _component_NodeList = __unplugin_components_0$1;
  const _component_n8n_button = resolveComponent("n8n-button");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass([
      _ctx.$style.card,
      _ctx.lastItem && _ctx.$style.last,
      _ctx.firstItem && _ctx.$style.first,
      !_ctx.loading && _ctx.$style.loaded
    ]),
    "data-test-id": "template-card",
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.onCardClick && _ctx.onCardClick(...args))
  }, [
    _ctx.loading ? (openBlock(), createElementBlock("div", {
      key: 0,
      class: normalizeClass(_ctx.$style.loading)
    }, [
      createVNode(_component_n8n_loading, {
        rows: 2,
        "shrink-last": false,
        loading: _ctx.loading
      }, null, 8, ["loading"])
    ], 2)) : _ctx.workflow ? (openBlock(), createElementBlock("div", _hoisted_1$1, [
      createVNode(_component_n8n_heading, {
        bold: true,
        size: "small"
      }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.workflow.name), 1)
        ]),
        _: 1
      }),
      !_ctx.simpleView ? (openBlock(), createElementBlock("div", {
        key: 0,
        class: normalizeClass(_ctx.$style.content)
      }, [
        _ctx.workflow.totalViews ? (openBlock(), createElementBlock("span", _hoisted_2$1, [
          createVNode(_component_n8n_text, {
            size: "small",
            color: "text-light"
          }, {
            default: withCtx(() => [
              createVNode(_component_font_awesome_icon, { icon: "eye" }),
              createTextVNode(" " + toDisplayString(_ctx.abbreviateNumber(_ctx.workflow.totalViews)), 1)
            ]),
            _: 1
          })
        ])) : createCommentVNode("", true),
        _ctx.workflow.totalViews ? (openBlock(), createElementBlock("div", {
          key: 1,
          class: normalizeClass(_ctx.$style.line),
          textContent: "|"
        }, null, 2)) : createCommentVNode("", true),
        createVNode(_component_n8n_text, {
          size: "small",
          color: "text-light"
        }, {
          default: withCtx(() => [
            createVNode(_component_TimeAgo, {
              date: _ctx.workflow.createdAt
            }, null, 8, ["date"])
          ]),
          _: 1
        }),
        _ctx.workflow.user ? (openBlock(), createElementBlock("div", {
          key: 2,
          class: normalizeClass(_ctx.$style.line),
          textContent: "|"
        }, null, 2)) : createCommentVNode("", true),
        _ctx.workflow.user ? (openBlock(), createBlock(_component_n8n_text, {
          key: 3,
          size: "small",
          color: "text-light"
        }, {
          default: withCtx(() => [
            createTextVNode("By " + toDisplayString(_ctx.workflow.user.username), 1)
          ]),
          _: 1
        })) : createCommentVNode("", true)
      ], 2)) : createCommentVNode("", true)
    ])) : createCommentVNode("", true),
    !_ctx.loading && _ctx.workflow ? (openBlock(), createElementBlock("div", {
      key: 2,
      class: normalizeClass([_ctx.$style.nodesContainer, _ctx.useWorkflowButton && _ctx.$style.hideOnHover])
    }, [
      _ctx.workflow.nodes ? (openBlock(), createBlock(_component_NodeList, {
        key: 0,
        nodes: _ctx.workflow.nodes,
        limit: _ctx.nodesToBeShown,
        size: "md"
      }, null, 8, ["nodes", "limit"])) : createCommentVNode("", true)
    ], 2)) : createCommentVNode("", true),
    _ctx.useWorkflowButton ? (openBlock(), createElementBlock("div", {
      key: 3,
      class: normalizeClass(_ctx.$style.buttonContainer)
    }, [
      _ctx.useWorkflowButton ? (openBlock(), createBlock(_component_n8n_button, {
        key: 0,
        outline: "",
        label: "Use workflow",
        "data-test-id": "use-workflow-button",
        onClick: withModifiers(_ctx.onUseWorkflowClick, ["stop"])
      }, null, 8, ["onClick"])) : createCommentVNode("", true)
    ], 2)) : createCommentVNode("", true)
  ], 2);
}
const cssModules$1 = {
  "$style": style0$1
};
const __unplugin_components_0 = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["render", _sfc_render$1], ["__cssModules", cssModules$1]]);
const _sfc_main = defineComponent({
  name: "TemplateList",
  components: {
    TemplateCard: __unplugin_components_0
  },
  props: {
    infiniteScrollEnabled: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean
    },
    useWorkflowButton: {
      type: Boolean,
      default: false
    },
    workflows: {
      type: Array,
      default: () => []
    },
    totalWorkflows: {
      type: Number,
      default: 0
    },
    simpleView: {
      type: Boolean,
      default: false
    },
    totalCount: {
      type: Number,
      default: 0
    }
  },
  mounted() {
    if (this.infiniteScrollEnabled) {
      const content2 = document.getElementById("content");
      if (content2) {
        content2.addEventListener("scroll", this.onScroll);
      }
    }
  },
  beforeUnmount() {
    const content2 = document.getElementById("content");
    if (content2) {
      content2.removeEventListener("scroll", this.onScroll);
    }
  },
  methods: {
    onScroll() {
      const loaderRef = this.$refs.loader;
      if (!loaderRef || this.loading) {
        return;
      }
      const rect = loaderRef.getBoundingClientRect();
      const inView = rect.top >= 0 && rect.left >= 0 && rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) && rect.right <= (window.innerWidth || document.documentElement.clientWidth);
      if (inView) {
        this.$emit("loadMore");
      }
    },
    onCardClick(event, id) {
      this.$emit("openTemplate", { event, id });
    },
    onUseWorkflow(event, id) {
      this.$emit("useWorkflow", { event, id });
    }
  }
});
const header = "_header_1ivn7_1";
const workflowButton = "_workflowButton_1ivn7_5";
const button = "_button_1ivn7_5";
const nodes = "_nodes_1ivn7_8";
const style0 = {
  header,
  workflowButton,
  button,
  nodes
};
const _hoisted_1 = {
  key: 0,
  "data-test-id": "template-count-label"
};
const _hoisted_2 = ["textContent"];
const _hoisted_3 = {
  key: 0,
  ref: "loader"
};
const _hoisted_4 = {
  key: 1,
  "data-test-id": "templates-loading-container"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_TemplateCard = __unplugin_components_0;
  return _ctx.loading || _ctx.workflows.length ? (openBlock(), createElementBlock("div", {
    key: 0,
    class: normalizeClass(_ctx.$style.list)
  }, [
    !_ctx.simpleView ? (openBlock(), createElementBlock("div", {
      key: 0,
      class: normalizeClass(_ctx.$style.header)
    }, [
      createVNode(_component_n8n_heading, {
        bold: true,
        size: "medium",
        color: "text-light"
      }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("templates.workflows")) + " ", 1),
          _ctx.totalCount > 0 ? (openBlock(), createElementBlock("span", _hoisted_1, "(" + toDisplayString(_ctx.totalCount) + ")", 1)) : createCommentVNode("", true),
          !_ctx.loading && _ctx.totalWorkflows ? (openBlock(), createElementBlock("span", {
            key: 1,
            textContent: toDisplayString(`(${_ctx.totalWorkflows})`)
          }, null, 8, _hoisted_2)) : createCommentVNode("", true)
        ]),
        _: 1
      })
    ], 2)) : createCommentVNode("", true),
    createBaseVNode("div", {
      class: normalizeClass(_ctx.$style.container)
    }, [
      (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.workflows, (workflow, index) => {
        return openBlock(), createBlock(_component_TemplateCard, {
          key: workflow.id,
          workflow,
          "first-item": index === 0,
          "simple-view": _ctx.simpleView,
          "last-item": index === _ctx.workflows.length - 1 && !_ctx.loading,
          "use-workflow-button": _ctx.useWorkflowButton,
          onClick: (e) => _ctx.onCardClick(e, workflow.id),
          onUseWorkflow: (e) => _ctx.onUseWorkflow(e, workflow.id)
        }, null, 8, ["workflow", "first-item", "simple-view", "last-item", "use-workflow-button", "onClick", "onUseWorkflow"]);
      }), 128)),
      _ctx.infiniteScrollEnabled ? (openBlock(), createElementBlock("div", _hoisted_3, null, 512)) : createCommentVNode("", true),
      _ctx.loading ? (openBlock(), createElementBlock("div", _hoisted_4, [
        (openBlock(), createElementBlock(Fragment, null, renderList(4, (n) => {
          return createVNode(_component_TemplateCard, {
            key: "index-" + n,
            loading: true,
            "first-item": _ctx.workflows.length === 0 && n === 1,
            "last-item": n === 4
          }, null, 8, ["first-item", "last-item"]);
        }), 64))
      ])) : createCommentVNode("", true)
    ], 2)
  ], 2)) : createCommentVNode("", true);
}
const cssModules = {
  "$style": style0
};
const __unplugin_components_2 = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules]]);
export {
  __unplugin_components_2 as _,
  __unplugin_components_0$1 as a
};
