// n8n-editor-ui/dist/assets/TemplateDetails-BHYTgWe7.js
import { d as defineComponent, c as openBlock, h as createElementBlock, j as createBaseVNode, i as createVNode, w as withCtx, k as createTextVNode, t as toDisplayString, n as normalizeClass, q as renderSlot, l as resolveComponent, _ as _export_sfc, d3 as _sfc_main$2, hR as _sfc_main$3, e0 as mapStores, aC as useTemplatesStore, hS as abbreviateNumber, hT as filterTemplateNodes, e as createBlock, F as Fragment, z as renderList, f as createCommentVNode } from "./index-TQ22MZub.js";
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  __name: "TemplateDetailsBlock",
  props: {
    title: {}
  },
  setup(__props) {
    return (_ctx, _cache) => {
      const _component_n8n_heading = resolveComponent("n8n-heading");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.block)
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.header)
        }, [
          createVNode(_component_n8n_heading, {
            tag: "h3",
            size: "small",
            color: "text-base"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.title), 1)
            ]),
            _: 1
          })
        ], 2),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.content)
        }, [
          renderSlot(_ctx.$slots, "default")
        ], 2)
      ], 2);
    };
  }
});
const block = "_block_czs8t_1";
const header = "_header_czs8t_5";
const content = "_content_czs8t_10";
const style0$1 = {
  block,
  header,
  content
};
const cssModules$1 = {
  "$style": style0$1
};
const __unplugin_components_1$1 = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["__cssModules", cssModules$1]]);
function isTemplatesWorkflow(template) {
  return !!template && "totalViews" in template;
}
function isFullTemplatesCollection(template) {
  return !!template && "description" in template && "categories" in template;
}
const _sfc_main = defineComponent({
  name: "TemplateDetails",
  components: {
    NodeIcon: _sfc_main$2,
    TemplateDetailsBlock: __unplugin_components_1$1,
    TimeAgo: _sfc_main$3
  },
  props: {
    template: {
      type: Object,
      required: true
    },
    blockTitle: {
      type: String,
      required: true
    },
    loading: {
      type: Boolean
    }
  },
  computed: {
    ...mapStores(useTemplatesStore)
  },
  methods: {
    abbreviateNumber,
    filterTemplateNodes,
    redirectToCategory(id) {
      this.templatesStore.resetSessionId();
      void this.$router.push(`/templates?categories=${id}`);
    },
    redirectToSearchPage(node) {
      this.templatesStore.resetSessionId();
      void this.$router.push(`/templates?search=${node.displayName}`);
    },
    isFullTemplatesCollection,
    isTemplatesWorkflow
  }
});
const icons = "_icons_1yw74_1";
const icon = "_icon_1yw74_1";
const text = "_text_1yw74_12";
const style0 = {
  icons,
  icon,
  text
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_n8n_loading = resolveComponent("n8n-loading");
  const _component_NodeIcon = _sfc_main$2;
  const _component_TemplateDetailsBlock = __unplugin_components_1$1;
  const _component_n8n_tags = resolveComponent("n8n-tags");
  const _component_TimeAgo = _sfc_main$3;
  const _component_n8n_text = resolveComponent("n8n-text");
  return openBlock(), createElementBlock("div", null, [
    createVNode(_component_n8n_loading, {
      loading: _ctx.loading,
      rows: 5,
      variant: "p"
    }, null, 8, ["loading"]),
    !_ctx.loading && _ctx.template && _ctx.template.nodes.length > 0 ? (openBlock(), createBlock(_component_TemplateDetailsBlock, {
      key: 0,
      title: _ctx.blockTitle
    }, {
      default: withCtx(() => [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.icons)
        }, [
          (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.filterTemplateNodes(_ctx.template.nodes), (node) => {
            return openBlock(), createElementBlock("div", {
              key: node.name,
              class: normalizeClass(_ctx.$style.icon)
            }, [
              createVNode(_component_NodeIcon, {
                "node-type": node,
                size: 24,
                "show-tooltip": true,
                onClick: ($event) => _ctx.redirectToSearchPage(node)
              }, null, 8, ["node-type", "onClick"])
            ], 2);
          }), 128))
        ], 2)
      ]),
      _: 1
    }, 8, ["title"])) : createCommentVNode("", true),
    !_ctx.loading && _ctx.isFullTemplatesCollection(_ctx.template) && _ctx.template.categories.length > 0 ? (openBlock(), createBlock(_component_TemplateDetailsBlock, {
      key: 1,
      title: _ctx.$locale.baseText("template.details.categories")
    }, {
      default: withCtx(() => [
        createVNode(_component_n8n_tags, {
          tags: _ctx.template.categories,
          "onClick:tag": _ctx.redirectToCategory
        }, null, 8, ["tags", "onClick:tag"])
      ]),
      _: 1
    }, 8, ["title"])) : createCommentVNode("", true),
    !_ctx.loading && _ctx.template ? (openBlock(), createBlock(_component_TemplateDetailsBlock, {
      key: 2,
      title: _ctx.$locale.baseText("template.details.details")
    }, {
      default: withCtx(() => [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.text)
        }, [
          _ctx.isTemplatesWorkflow(_ctx.template) ? (openBlock(), createBlock(_component_n8n_text, {
            key: 0,
            size: "small",
            color: "text-base"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("template.details.created")) + " ", 1),
              createVNode(_component_TimeAgo, {
                date: _ctx.template.createdAt
              }, null, 8, ["date"]),
              createTextVNode(" " + toDisplayString(_ctx.$locale.baseText("template.details.by")) + " " + toDisplayString(_ctx.template.user ? _ctx.template.user.username : "n8n team"), 1)
            ]),
            _: 1
          })) : createCommentVNode("", true)
        ], 2),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.text)
        }, [
          _ctx.isTemplatesWorkflow(_ctx.template) && _ctx.template.totalViews !== 0 ? (openBlock(), createBlock(_component_n8n_text, {
            key: 0,
            size: "small",
            color: "text-base"
          }, {
            default: withCtx(() => [
              createTextVNode(toDisplayString(_ctx.$locale.baseText("template.details.viewed")) + " " + toDisplayString(_ctx.abbreviateNumber(_ctx.template.totalViews)) + " " + toDisplayString(_ctx.$locale.baseText("template.details.times")), 1)
            ]),
            _: 1
          })) : createCommentVNode("", true)
        ], 2)
      ]),
      _: 1
    }, 8, ["title"])) : createCommentVNode("", true)
  ]);
}
const cssModules = {
  "$style": style0
};
const __unplugin_components_1 = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__cssModules", cssModules]]);
export {
  __unplugin_components_1 as _,
  isFullTemplatesCollection as i
};
