// n8n-editor-ui/dist/assets/WorkerView-Cr7HLqrl.js
import { _ as __unplugin_components_0 } from "./PushConnectionTracker-CYr4h_PG.js";
import { d as defineComponent, r as ref, c as openBlock, h as createElementBlock, j as createBaseVNode, i as createVNode, w as withCtx, q as renderSlot, n as normalizeClass, f as createCommentVNode, l as resolveComponent, _ as _export_sfc, e as createBlock, k as createTextVNode, t as toDisplayString, F as Fragment, z as renderList, d9 as useClipboard, a as useToast, g as useI18n, ci as shallowRef, o as onMounted, y as onBeforeUnmount, H as watch, co as toRaw, bt as h, bF as isProxy, iD as Chart$1, A as unref, p as computed, b as useRouter, a3 as useDocumentTitle, e0 as mapStores, K as useUIStore, Z as useRootStore, m as useSettingsStore } from "./index-TQ22MZub.js";
import { a as useOrchestrationStore, W as WORKER_HISTORY_LENGTH, u as usePushConnection } from "./usePushConnection-Bg_ZQZs0.js";
import { u as usePushConnectionStore } from "./pushConnection.store-DBI9rig9.js";
function averageWorkerLoadFromLoads(loads) {
  return loads.reduce((prev, curr) => prev + curr, 0) / loads.length;
}
function averageWorkerLoadFromLoadsAsString(loads) {
  return averageWorkerLoadFromLoads(loads).toFixed(2);
}
function memAsGb(mem) {
  return mem / 1024 / 1024 / 1024;
}
const _sfc_main$6 = /* @__PURE__ */ defineComponent({
  __name: "WorkerAccordion.ee",
  props: {
    icon: { default: "tasks" },
    iconColor: { default: "black" },
    initialExpanded: { type: Boolean, default: true }
  },
  setup(__props) {
    const props = __props;
    const expanded2 = ref(props.initialExpanded);
    function toggle() {
      expanded2.value = !expanded2.value;
    }
    return (_ctx, _cache) => {
      const _component_n8n_icon = resolveComponent("n8n-icon");
      const _component_n8n_text = resolveComponent("n8n-text");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(["accordion", _ctx.$style.container])
      }, [
        createBaseVNode("div", {
          class: normalizeClass({ [_ctx.$style.header]: true, [_ctx.$style.expanded]: expanded2.value }),
          onClick: toggle
        }, [
          createVNode(_component_n8n_icon, {
            icon: _ctx.icon,
            color: _ctx.iconColor,
            size: "small",
            class: "mr-2xs"
          }, null, 8, ["icon", "color"]),
          createVNode(_component_n8n_text, {
            class: normalizeClass(_ctx.$style.headerText),
            color: "text-base",
            size: "small",
            align: "left",
            bold: ""
          }, {
            default: withCtx(() => [
              renderSlot(_ctx.$slots, "title")
            ]),
            _: 3
          }, 8, ["class"]),
          createVNode(_component_n8n_icon, {
            icon: expanded2.value ? "chevron-up" : "chevron-down",
            bold: ""
          }, null, 8, ["icon"])
        ], 2),
        expanded2.value ? (openBlock(), createElementBlock("div", {
          key: 0,
          class: normalizeClass({ [_ctx.$style.description]: true, [_ctx.$style.collapsed]: !expanded2.value })
        }, [
          renderSlot(_ctx.$slots, "content")
        ], 2)) : createCommentVNode("", true)
      ], 2);
    };
  }
});
const container$1 = "_container_odamb_1";
const header = "_header_odamb_5";
const headerText = "_headerText_odamb_11";
const expanded = "_expanded_odamb_15";
const description = "_description_odamb_19";
const style0$6 = {
  container: container$1,
  header,
  headerText,
  expanded,
  description
};
const cssModules$6 = {
  "$style": style0$6
};
const WorkerAccordion = /* @__PURE__ */ _export_sfc(_sfc_main$6, [["__cssModules", cssModules$6]]);
const _hoisted_1$3 = ["href"];
const _hoisted_2$1 = ["href"];
const _sfc_main$5 = /* @__PURE__ */ defineComponent({
  __name: "WorkerJobAccordion.ee",
  props: {
    items: {}
  },
  setup(__props) {
    const props = __props;
    function runningSince(started) {
      let seconds = Math.floor(((/* @__PURE__ */ new Date()).getTime() - started.getTime()) / 1e3);
      const hrs = Math.floor(seconds / 3600);
      seconds -= hrs * 3600;
      const mnts = Math.floor(seconds / 60);
      seconds -= mnts * 60;
      return `${hrs}h ${mnts}m ${Math.floor(seconds)}s`;
    }
    return (_ctx, _cache) => {
      const _component_n8n_text = resolveComponent("n8n-text");
      return openBlock(), createBlock(WorkerAccordion, {
        icon: "tasks",
        "icon-color": "black",
        "initial-expanded": true
      }, {
        title: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("workerList.item.jobListTitle")) + " (" + toDisplayString(_ctx.items.length) + ") ", 1)
        ]),
        content: withCtx(() => [
          props.items.length > 0 ? (openBlock(), createElementBlock("div", {
            key: 0,
            class: normalizeClass(_ctx.$style.accordionItems)
          }, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(props.items, (item) => {
              return openBlock(), createElementBlock("div", {
                key: item.executionId,
                class: normalizeClass(_ctx.$style.accordionItem)
              }, [
                createBaseVNode("a", {
                  href: "/workflow/" + item.workflowId + "/executions/" + item.executionId
                }, " Execution " + toDisplayString(item.executionId) + " - " + toDisplayString(item.workflowName), 9, _hoisted_1$3),
                createVNode(_component_n8n_text, {
                  color: "text-base",
                  size: "small",
                  align: "left"
                }, {
                  default: withCtx(() => {
                    var _a;
                    return [
                      createTextVNode(" | Started at: " + toDisplayString((_a = new Date(item.startedAt)) == null ? void 0 : _a.toLocaleTimeString()) + " | Running for " + toDisplayString(runningSince(new Date(item.startedAt))) + " " + toDisplayString(item.retryOf ? `| Retry of: ${item.retryOf}` : "") + " | ", 1)
                    ];
                  }),
                  _: 2
                }, 1024),
                createBaseVNode("a", {
                  target: "_blank",
                  href: "/workflow/" + item.workflowId
                }, " (Open workflow)", 8, _hoisted_2$1)
              ], 2);
            }), 128))
          ], 2)) : (openBlock(), createElementBlock("div", {
            key: 1,
            class: normalizeClass(_ctx.$style.accordionItems)
          }, [
            createBaseVNode("span", {
              class: normalizeClass(_ctx.$style.empty)
            }, toDisplayString(_ctx.$locale.baseText("workerList.item.jobList.empty")), 3)
          ], 2))
        ]),
        _: 1
      });
    };
  }
});
const accordionItems$2 = "_accordionItems_kt5sk_1";
const accordionItem$2 = "_accordionItem_kt5sk_1";
const empty = "_empty_kt5sk_14";
const style0$5 = {
  accordionItems: accordionItems$2,
  accordionItem: accordionItem$2,
  empty
};
const cssModules$5 = {
  "$style": style0$5
};
const WorkerJobAccordion = /* @__PURE__ */ _export_sfc(_sfc_main$5, [["__cssModules", cssModules$5]]);
const _hoisted_1$2 = ["onClick"];
const _sfc_main$4 = /* @__PURE__ */ defineComponent({
  __name: "WorkerNetAccordion.ee",
  props: {
    items: {}
  },
  setup(__props) {
    const props = __props;
    const i18n = useI18n();
    const clipboard = useClipboard();
    const { showMessage } = useToast();
    function onCopyToClipboard(content) {
      try {
        void clipboard.copy(content);
        showMessage({
          title: i18n.baseText("workerList.item.copyAddressToClipboard"),
          type: "success"
        });
      } catch {
      }
    }
    return (_ctx, _cache) => {
      return openBlock(), createBlock(WorkerAccordion, {
        icon: "tasks",
        "icon-color": "black",
        "initial-expanded": false
      }, {
        title: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("workerList.item.netListTitle")) + " (" + toDisplayString(_ctx.items.length) + ") ", 1)
        ]),
        content: withCtx(() => [
          props.items.length > 0 ? (openBlock(), createElementBlock("div", {
            key: 0,
            class: normalizeClass(_ctx.$style.accordionItems)
          }, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(props.items, (item) => {
              return openBlock(), createElementBlock("div", {
                key: item.address,
                class: normalizeClass(_ctx.$style.accordionItem),
                onClick: ($event) => onCopyToClipboard(item.address)
              }, [
                createTextVNode(toDisplayString(item.family) + ": ", 1),
                createBaseVNode("span", {
                  class: normalizeClass(_ctx.$style.clickable)
                }, toDisplayString(item.address), 3),
                createTextVNode(" " + toDisplayString(item.internal ? "(internal)" : ""), 1)
              ], 10, _hoisted_1$2);
            }), 128))
          ], 2)) : createCommentVNode("", true)
        ]),
        _: 1
      });
    };
  }
});
const accordionItems$1 = "_accordionItems_10xds_1";
const accordionItem$1 = "_accordionItem_10xds_1";
const clickable = "_clickable_10xds_15";
const style0$4 = {
  accordionItems: accordionItems$1,
  accordionItem: accordionItem$1,
  clickable
};
const cssModules$4 = {
  "$style": style0$4
};
const WorkerNetAccordion = /* @__PURE__ */ _export_sfc(_sfc_main$4, [["__cssModules", cssModules$4]]);
const CommonProps = {
  data: {
    type: Object,
    required: true
  },
  options: {
    type: Object,
    default: () => ({})
  },
  plugins: {
    type: Array,
    default: () => []
  },
  datasetIdKey: {
    type: String,
    default: "label"
  },
  updateMode: {
    type: String,
    default: void 0
  }
};
const Props = {
  type: {
    type: String,
    required: true
  },
  ...CommonProps
};
function toRawIfProxy(obj) {
  return isProxy(obj) ? toRaw(obj) : obj;
}
function cloneProxy(obj) {
  let src = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : obj;
  return isProxy(src) ? new Proxy(obj, {}) : obj;
}
function setOptions(chart2, nextOptions) {
  const options = chart2.options;
  if (options && nextOptions) {
    Object.assign(options, nextOptions);
  }
}
function setLabels(currentData, nextLabels) {
  currentData.labels = nextLabels;
}
function setDatasets(currentData, nextDatasets, datasetIdKey) {
  const addedDatasets = [];
  currentData.datasets = nextDatasets.map((nextDataset) => {
    const currentDataset = currentData.datasets.find((dataset) => dataset[datasetIdKey] === nextDataset[datasetIdKey]);
    if (!currentDataset || !nextDataset.data || addedDatasets.includes(currentDataset)) {
      return {
        ...nextDataset
      };
    }
    addedDatasets.push(currentDataset);
    Object.assign(currentDataset, nextDataset);
    return currentDataset;
  });
}
function cloneData(data, datasetIdKey) {
  const nextData = {
    labels: [],
    datasets: []
  };
  setLabels(nextData, data.labels);
  setDatasets(nextData, data.datasets, datasetIdKey);
  return nextData;
}
const Chart = defineComponent({
  props: Props,
  setup(props, param) {
    let { expose } = param;
    const canvasRef = ref(null);
    const chartRef = shallowRef(null);
    expose({
      chart: chartRef
    });
    const renderChart = () => {
      if (!canvasRef.value) return;
      const { type, data, options, plugins, datasetIdKey } = props;
      const clonedData = cloneData(data, datasetIdKey);
      const proxiedData = cloneProxy(clonedData, data);
      chartRef.value = new Chart$1(canvasRef.value, {
        type,
        data: proxiedData,
        options: {
          ...options
        },
        plugins
      });
    };
    const destroyChart = () => {
      const chart2 = toRaw(chartRef.value);
      if (chart2) {
        chart2.destroy();
        chartRef.value = null;
      }
    };
    const update = (chart2) => {
      chart2.update(props.updateMode);
    };
    onMounted(renderChart);
    onBeforeUnmount(destroyChart);
    watch([
      () => props.options,
      () => props.data
    ], (param2, param1) => {
      let [nextOptionsProxy, nextDataProxy] = param2, [prevOptionsProxy, prevDataProxy] = param1;
      const chart2 = toRaw(chartRef.value);
      if (!chart2) {
        return;
      }
      let shouldUpdate = false;
      if (nextOptionsProxy) {
        const nextOptions = toRawIfProxy(nextOptionsProxy);
        const prevOptions = toRawIfProxy(prevOptionsProxy);
        if (nextOptions && nextOptions !== prevOptions) {
          setOptions(chart2, nextOptions);
          shouldUpdate = true;
        }
      }
      if (nextDataProxy) {
        const nextLabels = toRawIfProxy(nextDataProxy.labels);
        const prevLabels = toRawIfProxy(prevDataProxy.labels);
        const nextDatasets = toRawIfProxy(nextDataProxy.datasets);
        const prevDatasets = toRawIfProxy(prevDataProxy.datasets);
        if (nextLabels !== prevLabels) {
          setLabels(chart2.config.data, nextLabels);
          shouldUpdate = true;
        }
        if (nextDatasets && nextDatasets !== prevDatasets) {
          setDatasets(chart2.config.data, nextDatasets, props.datasetIdKey);
          shouldUpdate = true;
        }
      }
      if (shouldUpdate) {
        update(chart2);
      }
    }, {
      deep: true
    });
    return () => {
      return h("canvas", {
        ref: canvasRef
      });
    };
  }
});
const _sfc_main$3 = /* @__PURE__ */ defineComponent({
  __name: "WorkerChartsAccordion.ee",
  props: {
    workerId: {}
  },
  setup(__props) {
    var _a, _b, _c;
    const props = __props;
    const blankDataSet = (label, color, prefill = 0) => ({
      datasets: [
        {
          label,
          backgroundColor: color,
          data: prefill ? Array(Math.min(WORKER_HISTORY_LENGTH, prefill)).fill(0) : []
        }
      ],
      labels: Array(Math.min(WORKER_HISTORY_LENGTH, prefill)).fill("")
    });
    const orchestrationStore = useOrchestrationStore();
    const chartRefJobs = ref(void 0);
    const chartRefCPU = ref(void 0);
    const chartRefMemory = ref(void 0);
    const optionsBase = () => ({
      responsive: true,
      maintainAspectRatio: true,
      scales: {
        y: {
          type: "linear",
          display: true,
          position: "left",
          min: 0,
          suggestedMax: 5
        }
      }
      // uncomment to disable animation
      // animation: {
      // 	duration: 0,
      // },
    });
    const optionsJobs = optionsBase();
    const optionsCPU = optionsBase();
    if ((_a = optionsCPU.scales) == null ? void 0 : _a.y) optionsCPU.scales.y.suggestedMax = 100;
    const maxMemory = memAsGb((_b = orchestrationStore.workers[props.workerId]) == null ? void 0 : _b.totalMem) ?? 1;
    const optionsMemory = optionsBase();
    if ((_c = optionsMemory.scales) == null ? void 0 : _c.y) optionsMemory.scales.y.suggestedMax = maxMemory;
    const dataJobs = ref(
      blankDataSet("Job Count", "rgb(255, 111, 92)", WORKER_HISTORY_LENGTH)
    );
    const dataCPU = ref(
      blankDataSet("Processor Usage", "rgb(19, 205, 103)", WORKER_HISTORY_LENGTH)
    );
    const dataMemory = ref(
      blankDataSet("Memory Usage", "rgb(244, 216, 174)", WORKER_HISTORY_LENGTH)
    );
    orchestrationStore.$onAction(({ name, store }) => {
      var _a2, _b2;
      if (name === "updateWorkerStatus") {
        const prefillCount = WORKER_HISTORY_LENGTH - (((_a2 = store.workersHistory[props.workerId]) == null ? void 0 : _a2.length) ?? 0);
        const newDataJobs = blankDataSet("Job Count", "rgb(255, 111, 92)", prefillCount);
        const newDataCPU = blankDataSet(
          "Processor Usage",
          "rgb(19, 205, 103)",
          prefillCount
        );
        const newDataMemory = blankDataSet(
          "Memory Usage",
          "rgb(244, 216, 174)",
          prefillCount
        );
        (_b2 = store.workersHistory[props.workerId]) == null ? void 0 : _b2.forEach((item) => {
          var _a3;
          newDataJobs.datasets[0].data.push(item.data.runningJobsSummary.length);
          (_a3 = newDataJobs.labels) == null ? void 0 : _a3.push(new Date(item.timestamp).toLocaleTimeString());
          newDataCPU.datasets[0].data.push(averageWorkerLoadFromLoads(item.data.loadAvg));
          newDataCPU.labels = newDataJobs.labels;
          newDataMemory.datasets[0].data.push(maxMemory - memAsGb(item.data.freeMem));
          newDataMemory.labels = newDataJobs.labels;
        });
        dataJobs.value = newDataJobs;
        dataCPU.value = newDataCPU;
        dataMemory.value = newDataMemory;
      }
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(WorkerAccordion, {
        icon: "tasks",
        "icon-color": "black",
        "initial-expanded": false
      }, {
        title: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("workerList.item.chartsTitle")), 1)
        ]),
        content: withCtx(() => [
          createBaseVNode("div", {
            class: normalizeClass(_ctx.$style.charts)
          }, [
            createVNode(unref(Chart), {
              ref_key: "chartRefJobs",
              ref: chartRefJobs,
              type: "line",
              data: dataJobs.value,
              options: unref(optionsJobs),
              class: normalizeClass(_ctx.$style.chart)
            }, null, 8, ["data", "options", "class"]),
            createVNode(unref(Chart), {
              ref_key: "chartRefCPU",
              ref: chartRefCPU,
              type: "line",
              data: dataCPU.value,
              options: unref(optionsCPU),
              class: normalizeClass(_ctx.$style.chart)
            }, null, 8, ["data", "options", "class"]),
            createVNode(unref(Chart), {
              ref_key: "chartRefMemory",
              ref: chartRefMemory,
              type: "line",
              data: dataMemory.value,
              options: unref(optionsMemory),
              class: normalizeClass(_ctx.$style.chart)
            }, null, 8, ["data", "options", "class"])
          ], 2)
        ]),
        _: 1
      });
    };
  }
});
const accordionItems = "_accordionItems_2yo1p_1";
const accordionItem = "_accordionItem_2yo1p_1";
const charts = "_charts_2yo1p_14";
const chart = "_chart_2yo1p_14";
const style0$3 = {
  accordionItems,
  accordionItem,
  charts,
  chart
};
const cssModules$3 = {
  "$style": style0$3
};
const WorkerChartsAccordion = /* @__PURE__ */ _export_sfc(_sfc_main$3, [["__cssModules", cssModules$3]]);
const _sfc_main$2 = /* @__PURE__ */ defineComponent({
  __name: "WorkerCard.ee",
  props: {
    workerId: {}
  },
  setup(__props) {
    let interval;
    const orchestrationStore = useOrchestrationStore();
    const props = __props;
    const secondsSinceLastUpdateString = ref("0");
    const stale2 = ref(false);
    const worker = computed(() => {
      return orchestrationStore.getWorkerStatus(props.workerId);
    });
    const sortedWorkerInterfaces = computed(
      () => {
        var _a;
        return ((_a = worker.value) == null ? void 0 : _a.interfaces.toSorted((a, b) => a.family.localeCompare(b.family))) ?? [];
      }
    );
    function upTime(seconds) {
      const days = Math.floor(seconds / (3600 * 24));
      seconds -= days * 3600 * 24;
      const hrs = Math.floor(seconds / 3600);
      seconds -= hrs * 3600;
      const mnts = Math.floor(seconds / 60);
      seconds -= mnts * 60;
      return `${days}d ${hrs}h ${mnts}m ${Math.floor(seconds)}s`;
    }
    onMounted(() => {
      interval = setInterval(() => {
        const lastUpdated = orchestrationStore.getWorkerLastUpdated(props.workerId);
        if (!lastUpdated) {
          return;
        }
        const secondsSinceLastUpdate = Math.ceil((Date.now() - lastUpdated) / 1e3);
        stale2.value = secondsSinceLastUpdate > 10;
        secondsSinceLastUpdateString.value = secondsSinceLastUpdate.toFixed(0);
      }, 500);
    });
    onBeforeUnmount(() => {
      clearInterval(interval);
    });
    return (_ctx, _cache) => {
      const _component_n8n_heading = resolveComponent("n8n-heading");
      const _component_n8n_text = resolveComponent("n8n-text");
      const _component_n8n_card = resolveComponent("n8n-card");
      return worker.value ? (openBlock(), createBlock(_component_n8n_card, {
        key: 0,
        class: normalizeClass(_ctx.$style.cardLink)
      }, {
        header: withCtx(() => [
          createVNode(_component_n8n_heading, {
            tag: "h2",
            bold: "",
            class: normalizeClass(stale2.value ? [_ctx.$style.cardHeading, _ctx.$style.stale] : [_ctx.$style.cardHeading]),
            "data-test-id": "worker-card-name"
          }, {
            default: withCtx(() => [
              createTextVNode(" Name: " + toDisplayString(worker.value.senderId) + " (" + toDisplayString(worker.value.hostname) + ") ", 1),
              _cache[0] || (_cache[0] = createBaseVNode("br", null, null, -1)),
              createTextVNode(" Average Load: " + toDisplayString(unref(averageWorkerLoadFromLoadsAsString)(worker.value.loadAvg ?? [0])) + " | Free Memory: " + toDisplayString(unref(memAsGb)(worker.value.freeMem).toFixed(2)) + "GB / " + toDisplayString(unref(memAsGb)(worker.value.totalMem).toFixed(2)) + "GB " + toDisplayString(stale2.value ? " (stale)" : ""), 1)
            ]),
            _: 1
          }, 8, ["class"])
        ]),
        append: withCtx(() => [
          createBaseVNode("div", {
            ref: "cardActions",
            class: normalizeClass(_ctx.$style.cardActions)
          }, null, 2)
        ]),
        default: withCtx(() => [
          createBaseVNode("div", {
            class: normalizeClass(_ctx.$style.cardDescription)
          }, [
            createVNode(_component_n8n_text, {
              color: "text-light",
              size: "small",
              class: normalizeClass(_ctx.$style.container)
            }, {
              default: withCtx(() => [
                createBaseVNode("span", null, toDisplayString(_ctx.$locale.baseText("workerList.item.lastUpdated")) + " " + toDisplayString(secondsSinceLastUpdateString.value) + "s ago | n8n-Version: " + toDisplayString(worker.value.version) + " | Architecture: " + toDisplayString(worker.value.arch) + " ( " + toDisplayString(worker.value.platform) + ") | Uptime: " + toDisplayString(upTime(worker.value.uptime)), 1),
                createVNode(WorkerJobAccordion, {
                  items: worker.value.runningJobsSummary
                }, null, 8, ["items"]),
                createVNode(WorkerNetAccordion, { items: sortedWorkerInterfaces.value }, null, 8, ["items"]),
                createVNode(WorkerChartsAccordion, {
                  "worker-id": worker.value.senderId
                }, null, 8, ["worker-id"])
              ]),
              _: 1
            }, 8, ["class"])
          ], 2)
        ]),
        _: 1
      }, 8, ["class"])) : createCommentVNode("", true);
    };
  }
});
const container = "_container_s5el6_1";
const cardLink = "_cardLink_s5el6_5";
const cardHeading = "_cardHeading_s5el6_15";
const stale = "_stale_s5el6_21";
const cardDescription = "_cardDescription_s5el6_25";
const cardActions = "_cardActions_s5el6_32";
const style0$2 = {
  container,
  cardLink,
  cardHeading,
  stale,
  cardDescription,
  cardActions
};
const cssModules$2 = {
  "$style": style0$2
};
const WorkerCard = /* @__PURE__ */ _export_sfc(_sfc_main$2, [["__cssModules", cssModules$2]]);
const _sfc_main$1 = defineComponent({
  name: "WorkerList",
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/naming-convention
  components: { PushConnectionTracker: __unplugin_components_0, WorkerCard },
  props: {
    autoRefreshEnabled: {
      type: Boolean,
      default: true
    }
  },
  setup() {
    const router = useRouter();
    const i18n = useI18n();
    const pushConnection = usePushConnection({ router });
    return {
      i18n,
      pushConnection,
      ...useToast(),
      documentTitle: useDocumentTitle()
    };
  },
  computed: {
    ...mapStores(useRootStore, useUIStore, usePushConnectionStore, useOrchestrationStore),
    combinedWorkers() {
      const returnData = [];
      for (const workerId in this.orchestrationManagerStore.workers) {
        returnData.push(this.orchestrationManagerStore.workers[workerId]);
      }
      return returnData;
    },
    initialStatusReceived() {
      return this.orchestrationManagerStore.initialStatusReceived;
    },
    workerIds() {
      return Object.keys(this.orchestrationManagerStore.workers);
    },
    pageTitle() {
      return this.i18n.baseText("workerList.pageTitle");
    }
  },
  mounted() {
    this.documentTitle.set(this.pageTitle);
    this.$telemetry.track("User viewed worker view", {
      instance_id: this.rootStore.instanceId
    });
  },
  beforeMount() {
    if (window.Cypress !== void 0) {
      return;
    }
    this.pushConnection.initialize();
    this.pushStore.pushConnect();
    this.orchestrationManagerStore.startWorkerStatusPolling();
  },
  beforeUnmount() {
    if (window.Cypress !== void 0) {
      return;
    }
    this.orchestrationManagerStore.stopWorkerStatusPolling();
    this.pushStore.pushDisconnect();
    this.pushConnection.terminate();
  },
  methods: {
    averageLoadAvg(loads) {
      return (loads.reduce((prev, curr) => prev + curr, 0) / loads.length).toFixed(2);
    },
    getStatus(payload) {
      if (payload.runningJobsSummary.length > 0) {
        return "running";
      } else {
        return "success";
      }
    },
    getRowClass(payload) {
      return [this.$style.execRow, this.$style[this.getStatus(payload)]].join(" ");
    }
  }
});
const workerListHeader = "_workerListHeader_12j1s_1";
const card = "_card_12j1s_8";
const tableLoader = "_tableLoader_12j1s_12";
const style0$1 = {
  workerListHeader,
  card,
  tableLoader
};
const _hoisted_1$1 = { key: 0 };
const _hoisted_2 = { key: 1 };
const _hoisted_3 = { key: 0 };
const _hoisted_4 = { key: 1 };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_PushConnectionTracker = __unplugin_components_0;
  const _component_n8n_heading = resolveComponent("n8n-heading");
  const _component_n8n_spinner = resolveComponent("n8n-spinner");
  const _component_WorkerCard = resolveComponent("WorkerCard");
  return openBlock(), createElementBlock("div", null, [
    createVNode(_component_PushConnectionTracker, { class: "actions" }),
    createBaseVNode("div", {
      class: normalizeClass(_ctx.$style.workerListHeader)
    }, [
      createVNode(_component_n8n_heading, {
        tag: "h1",
        size: "2xlarge"
      }, {
        default: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.pageTitle), 1)
        ]),
        _: 1
      })
    ], 2),
    !_ctx.initialStatusReceived ? (openBlock(), createElementBlock("div", _hoisted_1$1, [
      createVNode(_component_n8n_spinner)
    ])) : (openBlock(), createElementBlock("div", _hoisted_2, [
      _ctx.workerIds.length === 0 ? (openBlock(), createElementBlock("div", _hoisted_3, toDisplayString(_ctx.$locale.baseText("workerList.empty")), 1)) : (openBlock(), createElementBlock("div", _hoisted_4, [
        (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.workerIds, (workerId) => {
          return openBlock(), createElementBlock("div", {
            key: workerId,
            class: normalizeClass(_ctx.$style.card)
          }, [
            createVNode(_component_WorkerCard, {
              "worker-id": workerId,
              "data-test-id": "worker-card"
            }, null, 8, ["worker-id"])
          ], 2);
        }), 128))
      ]))
    ]))
  ]);
}
const cssModules$1 = {
  "$style": style0$1
};
const WorkerList = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["render", _sfc_render], ["__cssModules", cssModules$1]]);
const _hoisted_1 = ["href"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "WorkerView",
  setup(__props) {
    const settingsStore = useSettingsStore();
    const uiStore = useUIStore();
    const goToUpgrade = () => {
      void uiStore.goToUpgrade("worker-view", "upgrade-worker-view");
    };
    return (_ctx, _cache) => {
      const _component_n8n_action_box = resolveComponent("n8n-action-box");
      return unref(settingsStore).isQueueModeEnabled && unref(settingsStore).isWorkerViewAvailable ? (openBlock(), createBlock(WorkerList, {
        key: 0,
        "data-test-id": "worker-view-licensed"
      })) : (openBlock(), createBlock(_component_n8n_action_box, {
        key: 1,
        "data-test-id": "worker-view-unlicensed",
        class: normalizeClass(_ctx.$style.actionBox),
        description: _ctx.$locale.baseText("workerList.actionBox.description"),
        "button-text": _ctx.$locale.baseText("workerList.actionBox.buttonText"),
        "onClick:button": goToUpgrade
      }, {
        heading: withCtx(() => [
          createBaseVNode("span", null, toDisplayString(_ctx.$locale.baseText("workerList.actionBox.title")), 1)
        ]),
        description: withCtx(() => [
          createTextVNode(toDisplayString(_ctx.$locale.baseText("workerList.actionBox.description")) + " ", 1),
          createBaseVNode("a", {
            href: _ctx.$locale.baseText("workerList.docs.url"),
            target: "_blank"
          }, toDisplayString(_ctx.$locale.baseText("workerList.actionBox.description.link")), 9, _hoisted_1)
        ]),
        _: 1
      }, 8, ["class", "description", "button-text"]));
    };
  }
});
const actionBox = "_actionBox_13ul9_1";
const style0 = {
  actionBox
};
const cssModules = {
  "$style": style0
};
const WorkerView = /* @__PURE__ */ _export_sfc(_sfc_main, [["__cssModules", cssModules]]);
export {
  WorkerView as default
};
